diff --git a/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh b/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
index ef18773b..59ed48d0 100755
--- a/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
+++ b/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
@@ -56,10 +56,16 @@ start_dutserver(){
 stop_dut_components(){
 	stop_drvhlpr
 	stop_dutserver
+	service_index=`ubus-cli "Firewall.Service.?" | grep 22222 | awk -F. '{print $3}'`
+	if [ -n "${service_index}" ]
+	then
+		ubus call Firewall.Service.${service_index} _del
+	fi
 }
 start_dut_components(){
 	start_drvhlpr_dut_mode
 	start_dutserver
+	ubus call Firewall.Service _add '{"parameters":{"DestPort": 22222, "Enable": "true", "IPVersion": 4, "Interface": "Device.IP.Interface.3.", "Protocol": 6}}'
 }
 ##########################################################################################################################
 # Get start/stop/restart flag
diff --git a/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh b/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
index ef18773b..59ed48d0 100755
--- a/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
+++ b/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
@@ -56,10 +56,16 @@ start_dutserver(){
 stop_dut_components(){
 	stop_drvhlpr
 	stop_dutserver
+	service_index=`ubus-cli "Firewall.Service.?" | grep 22222 | awk -F. '{print $3}'`
+	if [ -n "${service_index}" ]
+	then
+		ubus call Firewall.Service.${service_index} _del
+	fi
 }
 start_dut_components(){
 	start_drvhlpr_dut_mode
 	start_dutserver
+	ubus call Firewall.Service _add '{"parameters":{"DestPort": 22222, "Enable": "true", "IPVersion": 4, "Interface": "Device.IP.Interface.3.", "Protocol": 6}}'
 }
 ##########################################################################################################################
 # Get start/stop/restart flag
