diff -Naur a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2025-09-22 10:15:07.371910100 +0530
+++ b/CMakeLists.txt	2025-09-22 10:16:11.550118442 +0530
@@ -34,6 +34,7 @@
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 set(SOFTWARE_VER "8.2.0")
+set(MXL_RV_VER "2.1")
 
 INCLUDE (CheckIncludeFiles)
 INCLUDE (CheckFunctionExists)
diff -Naur a/config.h.cmake b/config.h.cmake
--- a/config.h.cmake	2025-09-22 10:20:13.198429723 +0530
+++ b/config.h.cmake	2025-09-22 10:20:37.145655625 +0530
@@ -2,6 +2,7 @@
 #define CONFIG_H
 
 #define SOFTWARE_VER "@SOFTWARE_VER@"
+#define MXL_RV_VER "@MXL_RV_VER@"
 #cmakedefine HAVE_GETIFADDRS
 #cmakedefine HAVE_LIBSSL
 #cmakedefine HAVE_LIBZ
diff -Naur a/udpst.c b/udpst.c
--- a/udpst.c	2025-09-22 10:22:28.046079709 +0530
+++ b/udpst.c	2025-09-22 10:26:43.909883970 +0530
@@ -250,7 +250,11 @@
         // Print banner or initialize JSON output object
         //
         if (!conf.jsonOutput) {
+#ifdef MXL_OPT
+                var = sprintf(scratch, SOFTWARE_TITLE "\nSoftware Ver: %s MxL Rev Ver: %s", SOFTWARE_VER, MXL_RV_VER);
+#else
                 var = sprintf(scratch, SOFTWARE_TITLE "\nSoftware Ver: %s", SOFTWARE_VER);
+#endif
                 if (repo.isServer)
                         var += sprintf(&scratch[var], ", Protocol Ver: %d-%d", PROTOCOL_MIN, PROTOCOL_VER);
                 else
