--- a/udpst_data.c	2026-02-22 19:03:58.455141274 -0800
+++ b/udpst_data.c	2026-02-22 19:05:46.497192528 -0800
@@ -355,6 +355,9 @@
                         totalsize += uvar;
                         nextsndbuf += payload;
                 }
+		if (totalsize/uvar > UDP_MAX_SEGMENTS) {
+			uvar = payload;
+		}
                 // update initial pdu info...
                 _gso_populate_header(lHdr, c, rttrd, totalsize, c->lpduSeqNo, uvar);    
                 //
