diff -Naur scripts/sbin/wifi scripts/sbin/wifi
--- 1/files/scripts/sbin/wifi	2019-06-26 15:05:46.284460376 +0300
+++ 2/files/scripts/sbin/wifi	2019-06-26 14:31:20.599078268 +0300
@@ -106,6 +106,37 @@
 	config_set "$device" hwmode "$hwmode"
 }
 
+#NOTE: this is a temporary solution till it will be supported in the UGW WLAN SL!!!
+update_wifi_ppa() {
+        local section
+        local ifname
+        local disable
+        local status="$1"
+
+        config_cb() {
+                config_get TYPE "$CONFIG_SECTION" TYPE
+                case "$TYPE" in
+                        wifi-iface)
+                                config_get device "$CONFIG_SECTION" device
+                                config_get disable "$device" disabled
+                                if [ $disable = "0" ]; then
+                                        append VIFS "$CONFIG_SECTION"
+                                fi
+                        ;;
+                esac
+        }
+        config_load "${cfgfile:-wireless}"
+
+	for section in $VIFS; do (
+                config_get ifname "$section" ifname
+                if [ "$status" = "enable" ]; then
+                        ppacmd addlan -i $ifname
+                else
+                        ppacmd dellan -i $ifname
+                fi
+        ); done
+}
+
 _wifi_updown() {
 	for device in ${2:-$DEVICES}; do (
 		config_get disabled "$device" disabled
@@ -121,6 +152,11 @@
 			echo "$device($iftype): Interface type not supported"
 		fi
 	); done
+	
+	#NOTE: this is a temporary solution till it will be supported in the UGW WLAN SL!!!
+	if [ "$OS_NAME" = "UGW" ]; then
+		update_wifi_ppa "$1"
+	fi
 }
 
 wifi_updown() {
