diff --git a/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh b/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
index ef18773b..b4928e25 100755
--- a/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
+++ b/files/scripts/2021-01-09/etc/wave/scripts/load_dut.sh
@@ -56,8 +56,10 @@ start_dutserver(){
 stop_dut_components(){
 	stop_drvhlpr
 	stop_dutserver
+	ubus call firewalld notify_firewall_change '{ "rule1" : "iptables -D zone_lan_input -p tcp --dport 22222 -j ACCEPT" }'
 }
 start_dut_components(){
+	ubus call firewalld notify_firewall_change '{ "rule1" : "iptables -I zone_lan_input -p tcp --dport 22222 -j ACCEPT" }'
 	start_drvhlpr_dut_mode
 	start_dutserver
 }
diff --git a/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh b/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
index ef18773b..b4928e25 100755
--- a/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
+++ b/files/scripts/2022-02-20/etc/wave/scripts/load_dut.sh
@@ -56,8 +56,10 @@ start_dutserver(){
 stop_dut_components(){
 	stop_drvhlpr
 	stop_dutserver
+	ubus call firewalld notify_firewall_change '{ "rule1" : "iptables -D zone_lan_input -p tcp --dport 22222 -j ACCEPT" }'
 }
 start_dut_components(){
+	ubus call firewalld notify_firewall_change '{ "rule1" : "iptables -I zone_lan_input -p tcp --dport 22222 -j ACCEPT" }'
 	start_drvhlpr_dut_mode
 	start_dutserver
 }
