diff -Naur a/src/extmem/memrchr_s.c b/src/extmem/memrchr_s.c
--- a/src/extmem/memrchr_s.c	2024-08-30 16:46:39.162269676 +0530
+++ b/src/extmem/memrchr_s.c	2024-08-30 16:48:41.477119576 +0530
@@ -37,6 +37,7 @@
 
 #ifdef HAVE_MEMRCHR
 #ifndef _ISOC11_SOURCE
+#pragma GCC diagnostic ignored "-Wredundant-decls"
 extern void* memrchr(const void *, int, size_t);
 #endif
 #endif
diff -Naur a/src/misc/qsort_s.c b/src/misc/qsort_s.c
--- a/src/misc/qsort_s.c	2024-08-30 16:46:16.058484213 +0530
+++ b/src/misc/qsort_s.c	2024-08-30 16:48:34.313188742 +0530
@@ -123,6 +123,7 @@
  */
 
 typedef int (*cmpfun)(const void *, const void *, void *);
+static unsigned char tmp[256];
 #ifdef HAVE___BUILTIN_CTZ
 #define ntz(x) __builtin_ctz((x))
 #else
@@ -170,7 +171,6 @@
 
 static void cycle(size_t width, unsigned char* ar[], int n)
 {
-    unsigned char tmp[256];
     size_t l;
     int i;
 
