diff -Naur org/configure.in new/configure.in
--- org/configure.in
+++ new/configure.in
@@ -1,7 +1,7 @@
 
 AC_REVISION($Revision: 1.21 $)
 
-AC_INIT(IFX MEI CPE Driver, 1.9.3, , drv_mei_cpe)
+AC_INIT(IFX MEI CPE Driver, 1.9.3-pd1, , drv_mei_cpe)
 AC_CONFIG_SRCDIR(src/Makefile.am)
 AM_INIT_AUTOMAKE([tar-pax])
 
diff -Naur org/src/drv_mei_cpe_linux.c new/src/drv_mei_cpe_linux.c
--- org/src/drv_mei_cpe_linux.c
+++ new/src/drv_mei_cpe_linux.c
@@ -1564,12 +1564,13 @@
 */
 static void MEI_GetDevInfoProc(struct seq_file *s)
 {
+/*
    if (!capable(CAP_NET_ADMIN))
    {
       printk(KERN_ERR "Permission denied" MEI_DRV_CRLF);
       return;
    }
-
+*/
    seq_printf(s, "MaxDeviceNumber=%d\n",MEI_DFEX_ENTITIES);
    seq_printf(s, "LinesPerDevice=%d\n",MEI_DFE_INSTANCE_PER_ENTITY);
    seq_printf(s, "ChannelsPerLine=%d\n",MEI_DEVICE_CFG_VALUE_GET(ChannelsPerLine));
