diff --git a/configure.in b/configure.in
index 6a6aba7..7e63460 100644
--- a/configure.in
+++ b/configure.in
@@ -1,7 +1,7 @@
 
 AC_REVISION($Revision: 1.21 $)
 
-AC_INIT(IFX MEI CPE Driver, 1.9.3-pd1, , drv_mei_cpe)
+AC_INIT(IFX MEI CPE Driver, 1.9.3-pd2, , drv_mei_cpe)
 AC_CONFIG_SRCDIR(src/Makefile.am)
 AM_INIT_AUTOMAKE([tar-pax])
 
diff --git a/src/drv_mei_cpe_linux.c b/src/drv_mei_cpe_linux.c
index f54c0e3..956b531 100755
--- a/src/drv_mei_cpe_linux.c
+++ b/src/drv_mei_cpe_linux.c
@@ -71,6 +71,9 @@
 #include <net/sock.h>
 #endif
 
+#include <linux/notifier.h>
+#include <linux/reboot.h>
+
 /* add MEI CPE debug/printout part */
 #include "drv_mei_cpe_dbg.h"
 
@@ -2266,6 +2269,20 @@ static struct platform_driver ltq_dsl_cpe_mei_driver = {
    .remove = ltq_dsl_cpe_mei_remove,
 };
 
+static int ltq_dsl_cpe_mei_reboot_notifier(struct notifier_block *nb,
+                                           unsigned long action,
+                                           void *data)
+{
+   printk("MEI_DRV: System reboot ongoing..." MEI_DRV_CRLF);
+   MEI_driver_exit();
+   return NOTIFY_DONE;
+}
+
+static struct notifier_block ltq_dsl_cpe_mei_reboot_nb = {
+   .notifier_call = ltq_dsl_cpe_mei_reboot_notifier,
+   .priority = 1,
+};
+
 static int ltq_dsl_cpe_mei_probe(struct platform_device *pdev)
 {
    int ret;
@@ -4351,6 +4368,8 @@ static IFX_void_t MEI_DRVOS_RemoveOwnPciPlatformDevices(void)
 
 static int MEI_module_init (void)
 {
+   int ret = 0;
+
    printk(KERN_INFO "%s, (c) 2007-2016 Lantiq Deutschland GmbH\n", &MEI_WHATVERSION[4]);
 
 #ifdef CONFIG_DEVFS_FS
@@ -4391,15 +4410,20 @@ static int MEI_module_init (void)
    MEI_DEVICE_CFG_VALUE_SET(ChannelsPerLine, 1);
    MEI_DEVICE_CFG_VALUE_SET(DfeChanDevices, dev_registered);
 
-   return platform_driver_register(&ltq_dsl_cpe_mei_driver);
+   ret = platform_driver_register(&ltq_dsl_cpe_mei_driver);
+
+   register_reboot_notifier(&ltq_dsl_cpe_mei_reboot_nb);
+
+   return ret;
 }
 
 
 static void MEI_module_exit(void)
 {
-    MEI_driver_exit();
-    platform_driver_unregister(&ltq_dsl_cpe_mei_driver);
-    MEI_DRVOS_RemoveOwnPciPlatformDevices();
+   unregister_reboot_notifier(&ltq_dsl_cpe_mei_reboot_nb);
+   MEI_driver_exit();
+   platform_driver_unregister(&ltq_dsl_cpe_mei_driver);
+   MEI_DRVOS_RemoveOwnPciPlatformDevices();
 }
 
 
