--- a/Makefile.am
+++ b/Makefile.am
@@ -23,18 +23,18 @@ SUBDIRS = . testsuite
 
 LTLIBOBJS = @LTLIBOBJS@
 
-lib_LTLIBRARIES = libargp.la
+lib_LTLIBRARIES = libargp-shared.la
 noinst_PROGRAMS = argp-test
 noinst_HEADERS = argp.h argp-fmtstream.h argp-namefrob.h # argp-comp.h
 
 EXTRA_DIST = mempcpy.c strchrnul.c strndup.c Versions
 
 # Leaves out argp-fs-xinl.c and argp-xinl.c 
-libargp_la_SOURCES = argp-ba.c argp-eexst.c argp-fmtstream.c \
+libargp_shared_la_SOURCES = argp-ba.c argp-eexst.c argp-fmtstream.c \
 		    argp-help.c argp-parse.c argp-pv.c \
 		    argp-pvh.c
 
-libargp_la_LIBADD = $(LTLIBOBJS)
+libargp_shared_la_LIBADD = $(LTLIBOBJS)
 
-argp_test_LDADD = libargp.la
+argp_test_LDADD = libargp-shared.la
 
--- a/testsuite/Makefile.am
+++ b/testsuite/Makefile.am
@@ -5,7 +5,7 @@ TS_ALL = $(TS_PROGS) $(TS_SH)
 
 noinst_PROGRAMS = $(TS_PROGS) ex1 ex3 ex4
 
-LDADD = -L../.libs -largp
+LDADD = -L../.libs -largp-shared
 
 EXTRA_DIST = $(TS_SH) run-tests
 CLEANFILES = test.out
