From 53e7a7a617462626c6123647c17bd0dee57b1c05 Mon Sep 17 00:00:00 2001
From: fromans <fromans@maxlinear.com>
Date: Fri, 17 Nov 2023 12:46:12 +0200
Subject: [PATCH] Temporary: fix 32 bit compilation issues

---
 src/ap/ieee802_11_eht.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index fc02f1198..e62ae776b 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -1190,7 +1190,7 @@ u16 hostapd_process_ml_assoc_req(struct hostapd_data *hapd,
 		goto out;
 	}
 
-	wpa_printf(MSG_DEBUG, "MLD: expected_common_info_len=%lu",
+	wpa_printf(MSG_DEBUG, "MLD: expected_common_info_len=%zu",
 		   common_info_len);
 
 	if (sizeof(*ml) + common_info_len > ml_len) {
@@ -1360,7 +1360,7 @@ u16 hostapd_process_ml_assoc_req(struct hostapd_data *hapd,
 		ml_len -= sub_elem_len;
 
 		wpa_printf(MSG_DEBUG, "MLD: link ctrl=0x%x, " MACSTR
-			   ", nstr bitmap len=%lu",
+			   ", nstr bitmap len=%zu",
 			   control, MAC2STR(link_info->peer_addr),
 			   link_info->nstr_bitmap_len);
 
-- 
2.43.0

