From 4c1113b1aa0cd6811fe7d3e2b3d80be39031c887 Mon Sep 17 00:00:00 2001
From: fromans <fromans@maxlinear.com>
Date: Fri, 17 Nov 2023 13:27:39 +0200
Subject: [PATCH] WLANRTSYS-60081: Compatibility stub for ATF configuration

---
 hostapd/mxl_config.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index c8194c728..131923c01 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -112,6 +112,8 @@ int mxl_hostapd_config_fill (struct hostapd_config *conf,
 			return 0;
 		}
 		mxl_conf->process_radar_slave_events = val;
+	} else if (os_strcmp(buf, "atf_config_file") == 0) {
+		wpa_printf(MSG_INFO, "atf_config_file will be ignored as it is not supported");
 	} else if  (os_strcmp(buf, "background_cac") == 0) {
 		int val = atoi(pos);
 		if (!IS_VALID_BOOL_RANGE(val)) {
-- 
2.43.0

