From fd67378ceef0db02099f3dc207d719a910eef1ba Mon Sep 17 00:00:00 2001
From: Gal Savion <gsavion@maxlinear.com>
Date: Wed, 12 Jul 2023 18:14:14 +0300
Subject: [PATCH] WLANRTSYS-56837: Expand Tx power offset to (-12) dB

---
 src/ap/ap_config.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 18ccd171c..13617c87b 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1325,7 +1325,7 @@ struct hostapd_config {
 	int acs_unii4;
 #endif /* CONFIG_ACS */
 
-	int sPowerSelection; /* tx power: 12%:9 25%:6 50%:3 75%:1 100%:0 */
+	int sPowerSelection; /* tx power: 6%:12 12%:9 25%:6 50%:3 75%:1 100%:0 */
 	int *sCoCPower;
 	int sCoCPowerSize;
 	int *sCoCAutoCfg;
-- 
2.43.0

