From 5f99ff60cf77f683661c950ddb8a4923796a6984 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Mon, 24 Jul 2023 15:27:38 +0200
Subject: [PATCH] WLANRTSYS-57156: Integrate LOW ACK enhancement

---
 src/ap/drv_callbacks.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 958f8ba78..799808ad8 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -952,7 +952,17 @@ void hostapd_notif_disassoc(struct hostapd_data *hapd, const u8 *addr)
 
 void hostapd_event_sta_low_ack(struct hostapd_data *hapd, const u8 *addr)
 {
-	struct sta_info *sta = ap_get_sta(hapd, addr);
+	struct sta_info *sta = NULL;
+	struct hostapd_iface *iface = hapd->iface;
+	size_t i;
+
+	for (i = 0; i < iface->num_bss; i++) {
+		sta = ap_get_sta(iface->bss[i], addr);
+		if (sta) {
+			hapd = iface->bss[i];
+			break;
+		}
+	}
 
 	if (!sta || !hapd->conf->disassoc_low_ack || sta->agreed_to_steer)
 		return;
-- 
2.43.0

