From 1f36bbad19c9174976f770476c2f1764120f7c38 Mon Sep 17 00:00:00 2001
From: Dmitrijs Martinovs <dmartinovs@maxlinear.com>
Date: Mon, 11 Sep 2023 14:45:42 +0200
Subject: [PATCH] WLANRTSYS-58289: Fix hostapd_drv_vendor_cmd

---
 src/ap/ap_drv_ops.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index 390ddf72c..bf2abcee2 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -383,7 +383,7 @@ static inline int hostapd_drv_vendor_cmd(struct hostapd_data *hapd,
 					 enum nested_attr nested_attr_flag,
 					 struct wpabuf *buf)
 {
-	if (hapd->driver == NULL || hapd->driver->vendor_cmd == NULL)
+	if (hapd->driver == NULL || hapd->driver->vendor_cmd == NULL || hapd->drv_priv == NULL)
 		return -1;
 	return hapd->driver->vendor_cmd(hapd->drv_priv, vendor_id, subcmd, data,
 					data_len, nested_attr_flag, buf);
-- 
2.43.0

