From 076c014ce0064efac1281ea95c294b1a0dbc2be7 Mon Sep 17 00:00:00 2001
From: Dmitrijs Martinovs <dmartinovs@maxlinear.com>
Date: Wed, 13 Sep 2023 11:49:25 +0200
Subject: [PATCH] WLANRTSYS-58548: Race condition in ctrl interface init

---
 hostapd/ctrl_iface.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index b79e1403f..e8becdced 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -7247,10 +7247,12 @@ fail:
 				   " allow connections - assuming it was left"
 				   "over from forced program termination");
 			if (unlink(fname) < 0) {
+				int error_num = errno;
 				wpa_printf(MSG_ERROR,
 					   "Could not unlink existing ctrl_iface socket '%s': %s",
-					   fname, strerror(errno));
-				goto fail;
+					   fname, strerror(error_num));
+				if (error_num != ENOENT) /* race condition between existence and access */
+					goto fail;
 			}
 			if (bind(s, (struct sockaddr *) &addr, sizeof(addr)) <
 			    0) {
@@ -8016,10 +8018,12 @@ fail:
 				   " allow connections - assuming it was left"
 				   "over from forced program termination");
 			if (unlink(fname) < 0) {
+				int error_num = errno;
 				wpa_printf(MSG_ERROR,
 					   "Could not unlink existing ctrl_iface socket '%s': %s",
-					   fname, strerror(errno));
-				goto fail;
+					   fname, strerror(error_num));
+				if (error_num != ENOENT) /* race condition between existence and access */
+					goto fail;
 			}
 			if (bind(s, (struct sockaddr *) &addr, sizeof(addr)) <
 			    0) {
-- 
2.43.0

