From 63dd13f1c76dec592e558e7d6a0154c8dcd81a15 Mon Sep 17 00:00:00 2001
From: Dmitrijs Martinovs <dmartinovs@maxlinear.com>
Date: Wed, 13 Sep 2023 14:48:17 +0200
Subject: [PATCH] WLANRTSYS-58550: Disabled VAPs should not handle probes

---
 src/ap/beacon.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index df3ca022c..8f33c2dff 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -1476,6 +1476,9 @@ void handle_probe_req(struct hostapd_data *hapd,
 	};
 #endif /* UBUS_SUPPORT */
 
+	if(!hapd->started || !hapd->beacon_set_done)
+		return;
+
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
 	    ssi_signal < hapd->iconf->rssi_ignore_probe_request)
 		return;
-- 
2.43.0

