From 5236638bb6a179225560c933a02f2b7ebe76a409 Mon Sep 17 00:00:00 2001
From: Nikita Chernikov <nchernikov@maxlinear.com>
Date: Thu, 7 Mar 2024 15:36:39 +0200
Subject: [PATCH] WLANRTSYS-63932 Build RNR element for 6G vaps also in probe
 response

Signed-off-by: Nikita Chernikov <nchernikov@maxlinear.com>
---
 src/ap/ieee802_11.c | 28 +++++++++++++++++++---------
 1 file changed, 19 insertions(+), 9 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index bb911d6ae..4b994f78c 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7866,7 +7866,7 @@ u8 * hostapd_eid_wb_chsw_wrapper(struct hostapd_data *hapd, u8 *eid)
 
 
 static size_t hostapd_eid_nr_db_len(struct hostapd_data *hapd,
-				    size_t *current_len)
+				    size_t *current_len, u32 type)
 {
 	struct hostapd_neighbor_entry *nr;
 	size_t total_len = 0, len = *current_len;
@@ -7876,6 +7876,11 @@ static size_t hostapd_eid_nr_db_len(struct hostapd_data *hapd,
 		if (!nr->nr || wpabuf_len(nr->nr) < 12)
 			continue;
 
+		/* Skip building RNR element from nr_db for non 6G VAPs in probe response */
+		if (!(is_6ghz_op_class(wpabuf_head_u8(nr->nr)[10])) &&
+			(type == WLAN_FC_STYPE_PROBE_RESP))
+			continue;
+
 		if (nr->short_ssid == hapd->conf->ssid.short_ssid)
 			continue;
 
@@ -8050,11 +8055,11 @@ size_t hostapd_eid_rnr_len(struct hostapd_data *hapd, u32 type)
 
 	switch (type) {
 	case WLAN_FC_STYPE_BEACON:
-		if (hapd->conf->rnr)
-			total_len += hostapd_eid_nr_db_len(hapd, &current_len);
 		/* fallthrough */
-
 	case WLAN_FC_STYPE_PROBE_RESP:
+		if (hapd->conf->rnr)
+			total_len += hostapd_eid_nr_db_len(hapd, &current_len, type);
+
 		if (mode == COLOCATED_LOWER_BAND || ap_mld)
 			total_len +=
 				hostapd_eid_rnr_multi_iface_len(hapd,
@@ -8083,7 +8088,7 @@ size_t hostapd_eid_rnr_len(struct hostapd_data *hapd, u32 type)
 
 
 static u8 * hostapd_eid_nr_db(struct hostapd_data *hapd, u8 *eid,
-			      size_t *current_len)
+			      size_t *current_len, u32 type)
 {
 	struct hostapd_neighbor_entry *nr;
 	size_t len = *current_len;
@@ -8094,6 +8099,11 @@ static u8 * hostapd_eid_nr_db(struct hostapd_data *hapd, u8 *eid,
 		if (!nr->nr || wpabuf_len(nr->nr) < 12)
 			continue;
 
+		/* Skip building RNR element from nr_db for non 6G VAPs in probe response */
+		if (!(is_6ghz_op_class(wpabuf_head_u8(nr->nr)[10])) &&
+			(type == WLAN_FC_STYPE_PROBE_RESP))
+			continue;
+
 		if (nr->short_ssid == hapd->conf->ssid.short_ssid)
 			continue;
 
@@ -8321,11 +8331,11 @@ u8 * hostapd_eid_rnr(struct hostapd_data *hapd, u8 *eid, u32 type)
 
 	switch (type) {
 	case WLAN_FC_STYPE_BEACON:
-		if (hapd->conf->rnr)
-			eid = hostapd_eid_nr_db(hapd, eid, &current_len);
 		/* fallthrough */
-
 	case WLAN_FC_STYPE_PROBE_RESP:
+		if (hapd->conf->rnr)
+			eid = hostapd_eid_nr_db(hapd, eid, &current_len, type);
+
 		if (mode == COLOCATED_LOWER_BAND || ap_mld)
 			eid = hostapd_eid_rnr_multi_iface(hapd, eid,
 							  &current_len);
@@ -8648,7 +8658,7 @@ u8 * hostapd_eid_mbssid(struct hostapd_data *hapd, u8 *eid, u8 *end,
 		cur_len = 0;
 
 		if (hapd->conf->rnr)
-			rnr_eid = hostapd_eid_nr_db(hapd, rnr_eid, &cur_len);
+			rnr_eid = hostapd_eid_nr_db(hapd, rnr_eid, &cur_len, frame_stype);
 		if (get_colocation_mode(hapd) == COLOCATED_LOWER_BAND)
 			rnr_eid = hostapd_eid_rnr_multi_iface(hapd, rnr_eid,
 							      &cur_len);
-- 
2.43.0

