--- a/cache.c
+++ b/cache.c
@@ -390,6 +390,7 @@ cache_dump_records(struct blob_buf *buf, const char *name)
 
 	last = avl_last_element(&records, last, avl);
 	for (r = avl_find_element(&records, name, r, avl); r; r = next) {
+		const struct dns_srv_data *dsd = (const struct dns_srv_data*)r->rdata;
 		switch (r->type) {
 		case TYPE_TXT:
 			if (r->txt && strlen(r->txt)) {
@@ -404,6 +405,8 @@ cache_dump_records(struct blob_buf *buf, const char *name)
 		case TYPE_SRV:
 			if (r->port)
 				blobmsg_add_u32(buf, "port", r->port);
+			if (r->rdlength > sizeof(*dsd))
+				blobmsg_add_string(buf, "target", (const char*)&dsd[1]);
 			break;
 
 		case TYPE_A:
