diff --git a/src/ra.c b/src/ra.c
index 01a8b72..9217b2a 100644
--- a/src/ra.c
+++ b/src/ra.c
@@ -176,8 +176,8 @@ int ra_init(const char *ifname, const struct in6_addr *ifid,
 		goto failure;
 
 	// Send RS
-	signal(SIGALRM, ra_send_rs);
-	ra_send_rs(SIGALRM);
+	//signal(SIGALRM, ra_send_rs);
+	//ra_send_rs(SIGALRM);
 
 	return 0;
 
@@ -191,7 +191,7 @@ failure:
 	return -1;
 }
 
-static void ra_send_rs(int signal __attribute__((unused)))
+__attribute__((unused)) static void ra_send_rs(int signal __attribute__((unused)))
 {
 	const struct sockaddr_in6 dest = {AF_INET6, 0, 0, ALL_IPV6_ROUTERS, if_index};
 	const struct icmpv6_opt llnull = {ND_OPT_SOURCE_LINKADDR, 1, {0}};
@@ -256,7 +256,7 @@ bool ra_link_up(void)
 		syslog(LOG_NOTICE, "carrier => %i event on %s", (int)!nocarrier, if_name);
 
 		rs_attempt = 0;
-		ra_send_rs(SIGALRM);
+		//ra_send_rs(SIGALRM);
 	}
 
 	return ret;
