From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maxlinear <nobody@maxlinear.com>
Date: Fri, 18 Feb 2022 09:51:50 +0100
Subject: Fix printing tc qdisc red flags

---
 tc/tc_red.c | 23 ++++-------------------
 1 file changed, 4 insertions(+), 19 deletions(-)

diff --git a/tc/tc_red.c b/tc/tc_red.c
index 88f5ff35..44e77d6d 100644
--- a/tc/tc_red.c
+++ b/tc/tc_red.c
@@ -102,23 +102,8 @@ int tc_red_eval_idle_damping(int Wlog, unsigned int avpkt, unsigned int bps, __u
 
 void tc_red_print_flags(__u32 flags)
 {
-	if (flags & TC_RED_ECN)
-		print_bool(PRINT_ANY, "ecn", "ecn ", true);
-	else
-		print_bool(PRINT_ANY, "ecn", NULL, false);
-
-	if (flags & TC_RED_HARDDROP)
-		print_bool(PRINT_ANY, "harddrop", "harddrop ", true);
-	else
-		print_bool(PRINT_ANY, "harddrop", NULL, false);
-
-	if (flags & TC_RED_ADAPTATIVE)
-		print_bool(PRINT_ANY, "adaptive", "adaptive ", true);
-	else
-		print_bool(PRINT_ANY, "adaptive", NULL, false);
-
-	if (flags & TC_RED_NODROP)
-		print_bool(PRINT_ANY, "nodrop", "nodrop ", true);
-	else
-		print_bool(PRINT_ANY, "nodrop", NULL, false);
+	print_bool(PRINT_ANY, "ecn", "ecn:%s ", flags & TC_RED_ECN);
+	print_bool(PRINT_ANY, "harddrop", "harddrop:%s ", flags & TC_RED_HARDDROP);
+	print_bool(PRINT_ANY, "adaptive", "adaptive:%s ", flags & TC_RED_ADAPTATIVE);
+	print_bool(PRINT_ANY, "nodrop", "nodrop:%s ", flags & TC_RED_NODROP);
 }
