From 6ef421c09f9fffd8c6403325f73153b8275f447d Mon Sep 17 00:00:00 2001
From: Hua Ma <hua.ma@linux.intel.com>
Date: Thu, 21 Jun 2018 17:38:25 +0800
Subject: [PATCH] Add support for xz wrapper changes

---
 fs/squashfs/xz_wrapper.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/fs/squashfs/xz_wrapper.c b/fs/squashfs/xz_wrapper.c
index 6bfaef73d065..9b9864a0aff9 100644
--- a/fs/squashfs/xz_wrapper.c
+++ b/fs/squashfs/xz_wrapper.c
@@ -39,10 +39,28 @@ struct squashfs_xz {
 	struct xz_buf buf;
 };
 
+/*
+ * These changes are needed for OpenWrt till and including version 15.5 CC
+ * because the default squashfs tool uses a different format. Upstream squashfs
+ * tools and the squashfs tools shipped with LEDE are using the default format.
+ * If this is not fixed, Squash over UBIFS will fail to mount.
+ * Use the GCC version to detect if we are in a OpenWrt CC environment or in
+ * something else, GCC version < 5 means we are in OpenWrt CC.
+ * TODO: try to get rid of this hack
+ */
+#if GCC_VERSION < 50000
+struct disk_comp_opts {
+	__le32 flags;
+	__le16 bit_opts;
+	__le16 fb;
+	__le32 dictionary_size;
+};
+#else
 struct disk_comp_opts {
 	__le32 dictionary_size;
 	__le32 flags;
 };
+#endif
 
 struct comp_opts {
 	int dict_size;
