From 449ce611dce64d10a1fe4529b182199c90eb9b4a Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Mon, 16 Jul 2018 11:07:10 +0200
Subject: [PATCH] PMCPESYS-51: thermal: ltq: Remove unused soc and name

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/thermal/ltq_thermal.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/drivers/thermal/ltq_thermal.c b/drivers/thermal/ltq_thermal.c
index e102e5d5927c..3b1bf8aaf18a 100644
--- a/drivers/thermal/ltq_thermal.c
+++ b/drivers/thermal/ltq_thermal.c
@@ -29,16 +29,9 @@
 
 #include  "thermal_core.h"
 
-enum ltq_soc {
-	LTQ_SOC_GRX500
-};
-
 struct ltq_thermal;
 
 struct ltq_thermal_tsens_data {
-	enum ltq_soc	soc;
-	char		*name;
-
 	void (*init)(struct platform_device *pdev, struct ltq_thermal *p);
 	int (*get_temp)(struct ltq_thermal *p);
 };
@@ -147,8 +140,6 @@ int ltq_grx500_get_temp(struct ltq_thermal *priv)
 
 /* Temperature sensor specific data */
 static struct ltq_thermal_tsens_data ltq_grx500_data = {
-	.soc		= LTQ_SOC_GRX500,
-	.name		= "TempSensor for GRX500",
 	.init		= ltq_grx500_init,
 	.get_temp	= ltq_grx500_get_temp,
 };
