From 71775fedc50defbf67b87ec316f38bca9b1ca8d0 Mon Sep 17 00:00:00 2001
From: "Belogaj, Dmitrij" <dmitrij.belogaj@intel.com>
Date: Fri, 3 Aug 2018 12:44:11 +0200
Subject: [PATCH] selftest:ptp: Make testptp.c compilable on machines without
 glibc

---
 tools/testing/selftests/ptp/testptp.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tools/testing/selftests/ptp/testptp.c b/tools/testing/selftests/ptp/testptp.c
index 5d2eae16f7ee..94dc235d9038 100644
--- a/tools/testing/selftests/ptp/testptp.c
+++ b/tools/testing/selftests/ptp/testptp.c
@@ -48,6 +48,7 @@
 #define CLOCK_INVALID -1
 #endif
 
+#if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
 /* clock_adjtime is not available in GLIBC < 2.14 */
 #if !__GLIBC_PREREQ(2, 14)
 #include <sys/syscall.h>
@@ -56,6 +57,7 @@ static int clock_adjtime(clockid_t id, struct timex *tx)
 	return syscall(__NR_clock_adjtime, id, tx);
 }
 #endif
+#endif
 
 static clockid_t get_clockid(int fd)
 {
