From c6e5bb97dadf8809cedb9744e5c30a23445d5d36 Mon Sep 17 00:00:00 2001
From: Wu Zhixian <zhixianx.wu@intel.com>
Date: Mon, 17 Sep 2018 13:38:23 +0800
Subject: [PATCH] fix comments

---
 drivers/watchdog/grx500_wdt.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index 536333e5a06c..ab1a283e31df 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -1,3 +1,4 @@
+/* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
  *                          (C) Copyright 2016~2017 Intel Corporation
@@ -402,12 +403,12 @@ static int grx500wdt_probe(struct platform_device *pdev)
 		/* Use enable_percpu_irq() for each Online CPU. */
 
 		preempt_disable();
-        if (cpu != smp_processor_id())
+		if (cpu != smp_processor_id())
 			smp_call_function_single(cpu,
 				(smp_call_func_t)Enable_WDT_intr,
 				(void *)&irq, 1);
-        else
-            Enable_WDT_intr((void *)&irq);
+		else
+			Enable_WDT_intr((void *)&irq);
 
 		preempt_enable();
 
