From f378dacfaddb2b14c07ddd804b50fa83c88bb529 Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Mon, 24 Sep 2018 16:55:25 +0200
Subject: [PATCH] PMCPESYS-89: spdmon: Remove old unused code

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/hwmon/spdmon.c | 58 --------------------------------------------------
 1 file changed, 58 deletions(-)

diff --git a/drivers/hwmon/spdmon.c b/drivers/hwmon/spdmon.c
index ce812006bbfe..8400c8e3304e 100644
--- a/drivers/hwmon/spdmon.c
+++ b/drivers/hwmon/spdmon.c
@@ -718,52 +718,6 @@ static void spd_mon_get_spd_data(void)
 }
 EXPORT_SYMBOL(spd_mon_get_spd_data);
 
-#if 0 /*I will keep this for later use*/
-static enum spd_class_t spd_mon_get_speed_class(struct spd_cnt *smcl)
-{
-	u32 lvt_spd, svt_spd;
-
-	spd_mon_get_spd_data();
-
-	/* Compare the values for all channels and found the lowest value */
-	if (spd_mon_data[0].lvt > spd_mon_data[1].lvt)
-		lvt_spd = spd_mon_data[1].lvt;
-	else
-		lvt_spd = spd_mon_data[0].lvt;
-
-	if (lvt_spd > spd_mon_data[2].lvt)
-		lvt_spd = spd_mon_data[2].lvt;
-
-	if (spd_mon_data[0].svt > spd_mon_data[1].svt)
-		svt_spd = spd_mon_data[1].svt;
-	else
-		svt_spd = spd_mon_data[0].svt;
-
-	if (svt_spd > spd_mon_data[2].svt)
-		svt_spd = spd_mon_data[2].svt;
-
-	/*Compare with threshold and Decide device class*/
-	/*Device is fast if both LVT and SVT class are fast; */
-	if ((lvt_spd > spd_lvt_f_low && lvt_spd < spd_lvt_f_high) &&
-	    (svt_spd > spd_svt_f_low && svt_spd < spd_svt_f_high))
-		return FAST; /* FAST DEVICE */
-
-	/*Device is Medium if LVT class is fast
-		or Med and SVT class is fast or Med;*/
-	if ((lvt_spd > spd_lvt_m_low && lvt_spd < spd_lvt_f_high) &&
-	    (svt_spd > spd_svt_m_low && svt_spd < spd_svt_f_high))
-		return MEDIUM; /* MEDIUM DEVICE */
-
-	if ((lvt_spd > spd_lvt_s_low && lvt_spd < spd_lvt_f_high) &&
-	    (svt_spd > spd_svt_s_low && svt_spd < spd_svt_f_high))
-		return SLOW; /* SLOW DEVICE */
-
-	return UNDEF;
-}
-EXPORT_SYMBOL(spd_mon_get_speed_class);
-#endif
-
-
 static enum spd_class_t spd_mon_get_speed_class(struct spd_cnt *smcl)
 {
 	int i;
@@ -809,18 +763,6 @@ static enum spd_class_t spd_mon_get_speed_class(struct spd_cnt *smcl)
 	if ((slow > fast) && (slow > med))
 		spd_mon_res = SLOW0;
 
-#if 0
-	/*now check this SMON value against the fused SMON value*/
-	smon_fused = spd_mon_get_fused_svt_spc();
-	if ((smon_fused == spd_mon_res) || (smon_fused == UNDEF))
-		spc = spd_mon_res;
-	else if ((smon_fused == SLOW0) || (spd_mon_res == SLOW0))
-		spc = SLOW0;
-	else if ((smon_fused == MEDIUM0) || (spd_mon_res == MEDIUM0))
-		spc = MEDIUM0;
-	else
-		spc = FAST0;
-#endif
 	/* we use only the fused values from now on */
 	smon_fused = spd_mon_get_fused_svt_spc();
 	if (smon_fused == UNDEF)
