From 3cc09e1259ee6686137469f08f507a14a8433c22 Mon Sep 17 00:00:00 2001
From: gsupriya <supriya1.g@intel.com>
Date: Mon, 8 Oct 2018 08:21:12 +0530
Subject: [PATCH] UGW_SW-31035:Adding Capability checks to ioctls and proc
 files

---
 net/mcast_helper/mcast_helper.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/net/mcast_helper/mcast_helper.c b/net/mcast_helper/mcast_helper.c
index 25c61d6f7d5f..4a45095cc3ec 100644
--- a/net/mcast_helper/mcast_helper.c
+++ b/net/mcast_helper/mcast_helper.c
@@ -1583,6 +1583,9 @@ static long mcast_helper_ioctl(struct file *f, unsigned int cmd, unsigned long a
 	struct net_device *upper_dev = NULL;
 	unsigned char s_mac[ETH_ALEN] = {0};
 
+	if (!capable(CAP_NET_ADMIN))
+		return -EPERM;
+
 	switch (cmd) {
 	case MCH_MEMBER_ENTRY_ADD:
 			if (copy_from_user(&mcast_mem, (MCAST_REC_t *)arg, sizeof(MCAST_REC_t))) {
@@ -1666,6 +1669,9 @@ int mcast_helper_seq_show(struct seq_file *seq, void *v)
 	MCAST_MAC_t *mac_rec = NULL;
 	struct list_head *gimc_list = mcast_helper_list_p(IPV4) ;
 
+	if (!capable(CAP_NET_ADMIN))
+		return -EPERM;
+
 	if (mch_acl_enabled) {
 		seq_printf(seq,
 				"%3s %10s "
@@ -1748,6 +1754,9 @@ int mcast_helper_seq_show6(struct seq_file *seq, void *v)
 	MCAST_MAC_t *mac_rec = NULL;
 	struct list_head *gimc_list = mcast_helper_list_p(IPV6) ;
 
+	if (!capable(CAP_NET_ADMIN))
+		return -EPERM;
+
 	if (mch_acl_enabled) {
 		seq_printf(seq,
 				"%3s %10s "
