From c14d5dc8f838e3a7e5e17e99facf7bb46bb12a42 Mon Sep 17 00:00:00 2001
From: Wu Zhixian <zhixianx.wu@intel.com>
Date: Thu, 18 Oct 2018 18:28:39 +0800
Subject: [PATCH] fix editor issue, change space to TAB

---
 drivers/dma/intel/cdma.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/drivers/dma/intel/cdma.c b/drivers/dma/intel/cdma.c
index 349d7c99fcea..1963b68c7432 100755
--- a/drivers/dma/intel/cdma.c
+++ b/drivers/dma/intel/cdma.c
@@ -533,10 +533,10 @@ int dma_device_register(_dma_device_info *dev)
 	case DMA_BURSTL_4DW:
 		txbl = IFX_DMA_BURSTL_4;
 		break;
-    case DMA_BURSTL_2DW:
-        txbl = IFX_DMA_BURSTL_2;
-        break;
-    default:
+	case DMA_BURSTL_2DW:
+		txbl = IFX_DMA_BURSTL_2;
+		break;
+	default:
 		txbl = DMA_PORT_DEFAULT_TX_BURST_LEN;
 	}
 	switch (dev->rx_burst_len) {
@@ -546,10 +546,10 @@ int dma_device_register(_dma_device_info *dev)
 	case DMA_BURSTL_4DW:
 		rxbl = IFX_DMA_BURSTL_4;
 		break;
-    case DMA_BURSTL_2DW:
-        rxbl = IFX_DMA_BURSTL_2;
-        break;
-    default:
+	case DMA_BURSTL_2DW:
+		rxbl = IFX_DMA_BURSTL_2;
+	break;
+	default:
 		rxbl = DMA_PORT_DEFAULT_RX_BURST_LEN;
 	}
 	spin_lock_irqsave(&g_dma_spinlock, flags);
