From a8c3bdf0361f506a6ff6b6729631225c01b85971 Mon Sep 17 00:00:00 2001
From: vijaysix <vijayx.singh@intel.com>
Date: Fri, 2 Nov 2018 15:25:07 +0530
Subject: [PATCH] UGW_SW-31838: address the review comment

---
 include/linux/extmark.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/linux/extmark.h b/include/linux/extmark.h
index dd0d1a1d070a..0944bc0306bd 100644
--- a/include/linux/extmark.h
+++ b/include/linux/extmark.h
@@ -35,7 +35,6 @@
 #define LROSESS_NO_OF_BITS 1
 #define LROSESS_MASK 0x80000000
 
-#endif
 
 #define GET_DATA_FROM_MARK_OPT(mark, mask, pos, value) \
 	(value = ((mark & mask) >> pos))
@@ -50,3 +49,5 @@
 	GET_DATA_FROM_MARK_OPT(mark, MASK(pos, len), pos, value)
 #define SET_DATA_FROM_MARK(mark, pos, len, value) \
 	SET_DATA_FROM_MARK_OPT(mark, MASK(pos, len), pos, value)
+
+#endif
