From bb05ee4e9fe9d59fb0feacd46bf2f85058d69ad2 Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Tue, 13 Nov 2018 15:25:35 +0100
Subject: [PATCH] PMCPESYS-104: regulator: tps65273: Fix minor coding style
 issues

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/regulator/tps65273-regulator.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/regulator/tps65273-regulator.c b/drivers/regulator/tps65273-regulator.c
index c267d5b47355..28abb4bdb534 100755
--- a/drivers/regulator/tps65273-regulator.c
+++ b/drivers/regulator/tps65273-regulator.c
@@ -66,11 +66,11 @@
 	.name		= "BUCK"#num,			\
 	.of_match	= NULL,				\
 	.regulators_node = of_match_ptr("regulators"),	\
-	.id		= TPS65273_BUCK##num##_ID,		\
+	.id		= TPS65273_BUCK##num##_ID,	\
 	.ops		= &tps65273_ops,		\
 	.type		= REGULATOR_VOLTAGE,		\
 	.owner		= THIS_MODULE,			\
-	.n_voltages     = ARRAY_SIZE(vout_vsel_table),	\
+	.n_voltages	= ARRAY_SIZE(vout_vsel_table),	\
 	.volt_table	= vout_vsel_table,		\
 	.vsel_reg	= TPS65273_REG_VOUT##num##_SEL,	\
 	.vsel_mask	= TPS65273_VOUT_SEL_MASK,	\
