From e90bb68151c1d0da24c5648250b3e116b5e33a93 Mon Sep 17 00:00:00 2001
From: ghoshp <purnendu.ghosh@intel.com>
Date: Fri, 1 Feb 2019 15:25:00 +0530
Subject: [PATCH] PONRTSYS-2771 fix for build failure due to signature mismatch

---
 include/net/ppa/ppa_api.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/net/ppa/ppa_api.h b/include/net/ppa/ppa_api.h
index 9cf2f77d1ab8..5e3f86cc6036 100644
--- a/include/net/ppa/ppa_api.h
+++ b/include/net/ppa/ppa_api.h
@@ -1000,7 +1000,7 @@ int32_t ppa_hook_get_bridge_status(uint32_t *f_enable, uint32_t flags);
 int32_t ppa_get_if_stats(PPA_IFNAME *, PPA_IF_STATS *, uint32_t);
 int32_t ppa_get_accel_stats(PPA_IFNAME *, PPA_ACCEL_STATS *, uint32_t);
 #if defined(PPA_IF_MIB) && PPA_IF_MIB
-int32_t ppa_get_netif_accel_stats(PPA_IFNAME *, PPA_NETIF_ACCEL_STATS *, uint32_t);
+int32_t ppa_get_netif_accel_stats(PPA_IFNAME *, PPA_NETIF_ACCEL_STATS *, uint32_t, struct rtnl_link_stats64 *);
 #endif
 int32_t ppa_hook_set_ppe_fastpath_enable(uint32_t f_enable, uint32_t flags);
 int32_t ppa_hook_get_ppe_fastpath_enable(uint32_t *f_enable, uint32_t flags);
