From 3be9cdf0cdc704ffbc5b294ba4b503221d5d3f52 Mon Sep 17 00:00:00 2001
From: Wu Zhixian <zhixianx.wu@intel.com>
Date: Thu, 16 May 2019 10:07:26 +0800
Subject: [PATCH] JIRA UGW_SW-37150:add shutdown for dc_ep

---
 drivers/net/ethernet/intel/vrx518/ep.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/ethernet/intel/vrx518/ep.c b/drivers/net/ethernet/intel/vrx518/ep.c
index 442be2ae24a1..40fc9d36299a 100755
--- a/drivers/net/ethernet/intel/vrx518/ep.c
+++ b/drivers/net/ethernet/intel/vrx518/ep.c
@@ -733,6 +733,7 @@ static struct pci_driver dc_ep_driver = {
 	.id_table = dc_ep_id_table,
 	.probe = dc_ep_probe,
 	.remove = dc_ep_remove,
+	.shutdown = dc_ep_remove,
 	/* PM not supported */
 	/* AER is controlled by RC */
 };
