From d9cdfba21aed3f8d90bd689cde0013ef1bb63d34 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Mon, 10 Jun 2019 10:37:21 +0800
Subject: [PATCH] Added show_state() API to dump the each and every process
 stack

---
 drivers/watchdog/grx500_wdt.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index fc94767861d7..b40ae0ce5586 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -284,6 +284,7 @@ static irqreturn_t grx500wdt_irq(int irqno, void *param)
 	struct watchdog_device *grx500_wdt;
 	struct grx500_wdt_struct *priv;
 	int i;
+	static int flag;
 
 	grx500_wdt = &per_cpu(grx500wdt, smp_processor_id());
 	priv = watchdog_get_drvdata(grx500_wdt);
@@ -301,6 +302,10 @@ static irqreturn_t grx500wdt_irq(int irqno, void *param)
 	WARN_ONCE(1, " IRQ %d triggered as WDT%d Timer Overflow on CPU %d!\n",
 		irqno, grx500_wdt->id, smp_processor_id());
 
+	if (flag == 0) {
+		show_state();
+		flag = 1;
+	}
 	return IRQ_HANDLED;
 }
 
