From c2cb0f03da262e047dcd95767d24eb09d385daba Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Mon, 17 Jun 2019 10:57:20 +0800
Subject: [PATCH] DRVLIB_SW-2398 - Add MT29F1G01ABA NAND flash into ltq_spinand

---
 drivers/mtd/ltq-spinand/ltq_spinand.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/mtd/ltq-spinand/ltq_spinand.c b/drivers/mtd/ltq-spinand/ltq_spinand.c
index 77f955078cc9..8076ceda7151 100644
--- a/drivers/mtd/ltq-spinand/ltq_spinand.c
+++ b/drivers/mtd/ltq-spinand/ltq_spinand.c
@@ -116,6 +116,8 @@ struct nand_flash_dev spinand_flash_ids[] = {
 	{"SPI NAND MT29F2G01AAAED 256MiB",
 		{ .id = {0x2c, 0x22} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 64},
+	{"SPI NAND MT29F1G01ABAFD 128MiB",
+		{ .id = {0x2c, 0x14} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 128},
 	{"SPI NAND MT29F2G01ABAGD 256MiB",
 		{ .id = {0x2c, 0x24} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 128},
