From 97b0790cd6de20fdb0af1028f01a745072f46fb4 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Fri, 5 Jul 2019 17:25:00 +0800
Subject: [PATCH] DRVLIB_SW-2520: Fix PRX300 GPIO driver compiler warnings

---
 drivers/gpio/gpio-intel-sso.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpio/gpio-intel-sso.c b/drivers/gpio/gpio-intel-sso.c
index 581ebf9418fa..5b26140297c0 100644
--- a/drivers/gpio/gpio-intel-sso.c
+++ b/drivers/gpio/gpio-intel-sso.c
@@ -139,6 +139,8 @@ sso_gpio_dir_out(struct gpio_chip *chip, unsigned int offset, int value)
 	sso_gpio_update_bit(priv->mmap, SSO_CPU, offset, value);
 	if (!priv->freq)
 		sso_gpio_update_bit(priv->mmap, SSO_CON0, SWU, 1);
+
+	return 0;
 }
 
 static int sso_gpio_get(struct gpio_chip *chip, unsigned int offset)
@@ -180,6 +182,7 @@ static int sso_gpio_freq_set(struct sso_gpio_priv *priv)
 
 	if (!priv->freq) {
 		us = US_SW;
+		off = US;
 	} else if (freq_idx < FPID_FREQ_RANK_MAX) {
 		mask = SSO_CON1_FPID_MASK;
 		off = SSO_CON1_FPID;
