From 57e6c8baeb89a22bdc9460ee39965315dabbc331 Mon Sep 17 00:00:00 2001
From: Deepansh Rastogi <deepansh.rastogi@intel.com>
Date: Fri, 6 Sep 2019 09:50:38 +0530
Subject: [PATCH] UGW_SW-41996: Adding new flag PPA_QOS_CL_CPU in uapi

---
 include/uapi/net/ppa_api.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index 43d38d6336da..517fcc65013c 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -409,6 +409,9 @@
 /*!
   \brief PPA_QOS_Q_F_WFQ_UPDATE */
 #define PPA_QOS_Q_F_WFQ_UPDATE                  0x00008000
+/*!
+  \brief PPA_QOS_CL_CPU */
+#define PPA_QOS_CL_CPU				0x00010000
 /*!
 	\brief PPA_F_PPPOATM */
 #define PPA_F_PPPOATM				0x80000000
