From fd2ad6f55a507d7987a29e1a73fbb5e800c59368 Mon Sep 17 00:00:00 2001
From: Rekha Eswaran <rekha.eswaran@intel.com>
Date: Wed, 30 Oct 2019 10:52:18 +0800
Subject: [PATCH] DRV_DPM_SW-74 - Fix DP swdev deadlock issue (PONRTSYS-5885)

---
 drivers/net/datapath/dpm/datapath_swdev.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 67c392815d68..2cc32be2f2ff 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -338,6 +338,7 @@ static int dp_swdev_filter_vlan(struct net_device *dev,
 			       br_info->br_device_name);
 			br_item->fid = br_info->fid;
 		} else {
+			swdev_unlock();
 			return -EOPNOTSUPP;
 		}
 		switchdev_trans_item_enqueue(trans, br_item,
@@ -345,9 +346,11 @@ static int dp_swdev_filter_vlan(struct net_device *dev,
 		swdev_unlock();
 		return 0;
 	}
+	swdev_lock();
 	br_item = switchdev_trans_item_dequeue(trans);
 	if (br_item)
 		dp_swdev_cfg_vlan(br_item, dev);
+	swdev_unlock();
 	return 0;
 }
 
