From 094a910160cd92701a87dafccc94b4940999bc0d Mon Sep 17 00:00:00 2001
From: Lishan Liu <lishan.liu@intel.com>
Date: Tue, 10 Dec 2019 17:56:37 +0800
Subject: [PATCH] TEP thermal to reconfig at every temperature read

---
 drivers/thermal/ltq_thermal.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/thermal/ltq_thermal.c b/drivers/thermal/ltq_thermal.c
index a898dcefc1a4..3e4575152318 100644
--- a/drivers/thermal/ltq_thermal.c
+++ b/drivers/thermal/ltq_thermal.c
@@ -47,7 +47,6 @@ struct ltq_thermal_sensor {
 	int temp;
 	int last_temp;
 	int emul_temp;
-
 	void *drvdata;
 };
 
@@ -116,9 +115,9 @@ static int ltq_prx300_get_temp(struct ltq_thermal_sensor *sensor)
 	if (!priv)
 		return ts_val;
 
-	/* Select a channel */
-	regmap_update_bits(priv->chiptop, CTRL_REG_ID(sensor->id),
-			   SOC_MASK | VSAMPLE_MASK, SOC_NC | VSMPLE_NV);
+	/* configure to temp channel */
+	regmap_write(priv->chiptop, CTRL_REG_ID(sensor->id),
+		     SOC_NC | VSMPLE_NV);
 
 	/* Converted data availability */
 	if (regmap_read_poll_timeout(priv->chiptop, DATA_REG_ID(sensor->id),
