From 0227021a145ec4c1fa4e759456bc66823d125e58 Mon Sep 17 00:00:00 2001
From: Alexander Abraham <alexander.abraham1@intel.com>
Date: Mon, 23 Dec 2019 17:01:32 +0530
Subject: [PATCH] UGW_SW-46333, UGW_SW-45661, add missing capablility checks in
 mtd ioctl

---
 drivers/mtd/mtdchar.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/mtd/mtdchar.c b/drivers/mtd/mtdchar.c
index b43b8edc18ec..7b3086ccc6a7 100644
--- a/drivers/mtd/mtdchar.c
+++ b/drivers/mtd/mtdchar.c
@@ -1096,6 +1096,9 @@ static long mtdchar_unlocked_ioctl(struct file *file, u_int cmd, u_long arg)
 {
 	int ret;
 
+	if (!capable(CAP_SYS_RESOURCE))
+		return -EPERM;
+
 	mutex_lock(&mtd_mutex);
 	ret = mtdchar_ioctl(file, cmd, arg);
 	mutex_unlock(&mtd_mutex);
@@ -1122,6 +1125,9 @@ static long mtdchar_compat_ioctl(struct file *file, unsigned int cmd,
 	void __user *argp = compat_ptr(arg);
 	int ret = 0;
 
+	if (!capable(CAP_SYS_RESOURCE))
+		return -EPERM;
+
 	mutex_lock(&mtd_mutex);
 
 	switch (cmd) {
