From dbe8b6452ee7cd91078a17459a67e6fbf46109d5 Mon Sep 17 00:00:00 2001
From: agaraix <anath.bandux.garai@intel.com>
Date: Tue, 4 Feb 2020 08:05:18 +0530
Subject: [PATCH] UGW_SW-46847 : net: dev: avoid extmark ifdef in xmit_one()

---
 net/core/dev.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index bbd0b8229d73..2a3b85b7b3d6 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -160,6 +160,12 @@ struct list_head ptype_base[PTYPE_HASH_SIZE] __read_mostly;
 struct list_head ptype_all __read_mostly;	/* Taps */
 static struct list_head offload_base __read_mostly;
 
+#ifdef CONFIG_NETWORK_EXTMARK
+#define match_extmark_filtertap(skb) (skb->extmark & FILTERTAP_MASK)
+#else
+#define match_extmark_filtertap(skb) 0
+#endif
+
 #if IS_ENABLED(CONFIG_MCAST_HELPER)
 int (*mcast_helper_sig_check_update_ptr)(struct sk_buff *skb) = NULL;
 EXPORT_SYMBOL(mcast_helper_sig_check_update_ptr);
@@ -2973,10 +2979,8 @@ static int xmit_one(struct sk_buff *skb, struct net_device *dev,
 	unsigned int len;
 	int rc;
 
-#ifdef CONFIG_NETWORK_EXTMARK
-	if (!(skb->extmark & FILTERTAP_MASK))
-#endif
-	if (!list_empty(&ptype_all) || !list_empty(&dev->ptype_all))
+	if (!match_extmark_filtertap(skb) &&
+	    (!list_empty(&ptype_all) || !list_empty(&dev->ptype_all)))
 		dev_queue_xmit_nit(skb, dev);
 
 	len = skb->len;
