From 2f51ee16086cf60e388e3ea05aba7c4d4d1a5fb0 Mon Sep 17 00:00:00 2001
From: "Vishnu Swaroop Kumar Sarma, Duddu" <duddu.swaroop@intel.com>
Date: Fri, 20 Mar 2020 00:20:04 +0530
Subject: [PATCH] UGW_SW-48467 regulator shutdown failure case recovery

---
 drivers/regulator/tps65273-regulator.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/regulator/tps65273-regulator.c b/drivers/regulator/tps65273-regulator.c
index ae989b525caa..3dee41770913 100755
--- a/drivers/regulator/tps65273-regulator.c
+++ b/drivers/regulator/tps65273-regulator.c
@@ -318,14 +318,21 @@ static void tps65273_shutdown(struct i2c_client *client)
 {
 	struct tps65273 *tps = i2c_get_clientdata(client);
 	int i, ret = 0;
+	int count = 5;
 
 	dev_dbg(&client->dev, "function %s is called\n", __func__);
 
 	for (i = 0; i < TPS65273_NUM_REGULATOR; i++) {
 		struct regulator_dev *rdev = tps->rdev[i];
-		ret = regmap_update_bits(tps->regmap,
+		do {
+			ret = regmap_update_bits(tps->regmap,
 					TPS65273_REG_VOUT1_SEL + i,
 					TPS65273_VOUT_SEL_GO_BIT, 0);
+			/* try out multiple times for success */
+			/* fail case may be signal with ERESTARTSYS */
+			if (ret == 0)
+				break;
+		} while(--count);
 		if (ret < 0) {
 			dev_err(&client->dev,
 				"shutdown regulator failed: (%d)\n", ret);
