From 2caa0d37c90564814ee0c26dc4f61bfaa72202d7 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Fri, 20 Mar 2020 11:19:30 +0800
Subject: [PATCH] PONRTSYS-5001: add regmap driver debugfs

---
 drivers/base/regmap/regmap-icc.c | 129 +++++++++------------------------------
 1 file changed, 28 insertions(+), 101 deletions(-)

diff --git a/drivers/base/regmap/regmap-icc.c b/drivers/base/regmap/regmap-icc.c
index c40eee74dc1c..f9cb0aebf23a 100644
--- a/drivers/base/regmap/regmap-icc.c
+++ b/drivers/base/regmap/regmap-icc.c
@@ -14,11 +14,6 @@
 
 #include "internal.h"
 
-enum{
-	REGMAP_DBG_READ,
-	REGMAP_DBG_WRITE,
-};
-
 struct regmap_icc_context {
 	void __iomem *regs;
 	phys_addr_t  paddr;
@@ -302,125 +297,57 @@ static ssize_t
 regmap_reg_dbg(struct file *s, const char __user *buffer,
 	       size_t count, loff_t *pos)
 {
-	char buf[128] = {0};
-	char *cact, *caddr, *cval, *cmask, *cclient;
-	u32 reg_addr, reg_val, action, client_id;
-	int i, len;
+	char buf[64] = {0};
+	char action;
+	u32 icc_id, reg, val = 0;
+	size_t buf_size;
+	int ret;
 
 	if (!capable(CAP_SYS_ADMIN))
 		return -EPERM;
 
-	len = count < sizeof(buf) ? count : sizeof(buf) - 1;
-	if (copy_from_user(buf, buffer, len))
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	buf_size = min(count, sizeof(buf) - 1);
+	if (copy_from_user(buf, buffer, buf_size))
 		return -EFAULT;
 
 	if (strcmp(buf, "help") == 0)
 		goto __err_help;
 
-	pr_info("buf: %s\n", buf);
-
-	caddr = NULL;
-	cval  = NULL;
-	cmask = NULL;
-	cclient = NULL;
-	cact = buf;
-	for (i = 0; buf[i] != ' ' && buf[i] != 0; i++)
-		;
-
-	if (buf[i] == 0)
-		goto __err_help;
-
-	buf[i++] = 0;
-
-	pr_info("cact: %s\n", cact);
-	if (!strcmp(cact, "r"))
-		action = REGMAP_DBG_READ;
-	else if (!strcmp(cact, "w"))
-		action = REGMAP_DBG_WRITE;
-	else
+	if (sscanf(buf, "%c %u 0x%x 0x%x", &action, &icc_id, &reg, &val) > 4)
 		goto __err_help;
 
-	cclient = &buf[i];
-	for (; buf[i] != ' ' && buf[i] != 0; i++)
-		;
-
-	if (buf[i] == 0)
-		goto __action;
-	buf[i++] = 0;
-
-	pr_info("client: %s\n", cclient);
-
-	caddr = &buf[i];
-	for (; buf[i] != ' ' && buf[i] != 0; i++)
-		;
-
-	if (buf[i] == 0)
-		goto __action;
-	buf[i++] = 0;
-
-	pr_info("addr: %s\n", caddr);
-
-	cval = &buf[i];
-	for (; buf[i] != ' ' && buf[i] != 0; i++)
-		;
-	if (buf[i] == 0)
-		goto __action;
-	buf[i++] = 0;
-	pr_info("cval: %s\n", cval);
-
-	cmask = &buf[i];
-	for (; buf[i] != ' ' && buf[i] != 0; i++)
-		;
-	buf[i++] = 0;
-	pr_info("cmask: %s\n", cmask);
-
-	if (!cclient || kstrtou32(cclient, 0, &client_id)) {
-		pr_err("Failed to parse reg value!\n");
-		goto __err_help;
-	}
 	/*its not allowed client */
-	if (!(client_id >= REGMAP_CGU && client_id <= REGMAP_CT)) {
+	if (!(icc_id >= REGMAP_CGU && icc_id <= REGMAP_CT)) {
 		pr_err("Possible values for client id is [%u] to [%u]!\n",
 				REGMAP_CGU, REGMAP_CT);
 		goto __err_help;
 	}
-__action:
-	switch (action) {
-	case REGMAP_DBG_WRITE:
-		if (!cval || kstrtou32(cval, 0, &reg_val)) {
-			pr_err("Failed to parse reg value!\n");
-			goto __err_help;
-		}
-	case REGMAP_DBG_READ:
-		if (!caddr || kstrtou32(caddr, 0, &reg_addr)) {
-			pr_err("Failed to parse reg address!\n");
-			goto __err_help;
-		}
-		break;
-	default:
-		break;
-	}
 
 	switch (action) {
-	case REGMAP_DBG_WRITE:
-		pr_info("write: regaddr: 0x%08x, regval: 0x%08x\n",
-			reg_addr, reg_val);
-		if (icc_regmap_sync_write(client_id, reg_addr, reg_val) != 0)
-			pr_err("icc_regmap_sync_write Failed\n");
-		break;
-	case REGMAP_DBG_READ:
-		pr_info("icc read addr: 0x%08x\n", reg_addr);
-		if (icc_regmap_sync_read(client_id, reg_addr, &reg_val) != 0)
-			pr_err("icc_regmap_sync_read Failed\n");
-
-		pr_info("read addr: 0x%08x, val: 0x%08x\n",
-			reg_addr, reg_val);
-		break;
+	case 'w':
+		ret = icc_regmap_sync_write(icc_id, reg, val);
+		if (ret != 0)
+			pr_info("icc_regmap_sync_write failed %d\n", ret);
+	break;
+	case 'r':
+		ret = icc_regmap_sync_read(icc_id, reg, &val);
+		if (ret != 0)
+			pr_info("icc_regmap_sync_read failed %d\n", ret);
+		else
+			pr_info("read: addr 0x%x val 0x%x\n", reg, val);
+	break;
+	default:
+		goto __err_help;
 	}
 
 	return count;
 
 __err_help:
+	pr_info("icc client ids: CGU %u RCU %u CHIPTOP %u\n",
+		REGMAP_CGU, REGMAP_RCU, REGMAP_CT);
 	pr_info("echo r <icc client> <addr> > dbg\n");
 	pr_info("echo w <icc client> <addr> <value> > dbg\n");
 	return count;
