From 7f84ad3af913db93a71bd0d903c5d2784b42e896 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Thu, 26 Mar 2020 02:24:32 +0200
Subject: [PATCH] PONRTSYS-7025: Align pp_api.h

Signed-off-by: Emergi Idan <idan.emergi@intel.com>
---
 include/linux/pp_api.h | 75 +++++++++++++++++++++++++++++++++++++-------------
 1 file changed, 56 insertions(+), 19 deletions(-)

diff --git a/include/linux/pp_api.h b/include/linux/pp_api.h
index 0aef645f0a4d..ab18d6f78d5a 100755
--- a/include/linux/pp_api.h
+++ b/include/linux/pp_api.h
@@ -13,7 +13,7 @@
 #include <linux/pp_buffer_mgr_api.h>
 #include <linux/pp_qos_api.h>
 
-#ifdef CONFIG_PHY_INTEL_COMBO_LGM
+#ifdef CONFIG_PPV4_LGM
 #include <linux/learning_layer_api.h>
 #endif
 
@@ -157,9 +157,18 @@ enum pp_version_type {
  * @brief pp decriptor packet user-define host structure
  */
 struct pp_pkt_ud {
+	u8  rx_port;              /*! rx port id                   */
+	u8  rsrv;                 /*! reserved                     */
+	u16 tdox_flow:10,         /*! turbo dox flow               */
+	    lro:1;                /*! LRO exception                */
+	u32 sess_id:24,           /*! session id                   */
+	    is_exc_sess:1;        /*! if exception session         */
+	u32 hash_sig;             /*! hash signature result        */
+	u32 hash_h1;              /*! hash H1 result               */
+	u32 hash_h2;              /*! hash H2 result               */
+#ifdef PP_FULL_DESC_DBG
 	u8  int_proto_info:4,     /*! internal protocol info       */
 	    ext_proto_info:4;     /*! external protocol info       */
-	u8  rx_port;              /*! rx port id                   */
 	u8  ttl_exp:1,            /*! TTL expired                  */
 	    ip_opt:1,             /*! IP/IPv6 option               */
 	    ext_df:1,             /*! external don't frag bit set  */
@@ -177,19 +186,13 @@ struct pp_pkt_ud {
 	u8  int_l4_off;           /*! internal L4 offset           */
 	u8  ext_frag_off;         /*! external fragment offset     */
 	u8  int_frag_off;         /*! internal fragment offset     */
-	u16 tdox_flow:10,         /*! turbo dox flow               */
-	    lro:1;                /*! LRO exception                */
 	u8  l2_off;               /*! L2 offset (pre-L2 size)      */
-	u32 sess_id:24,           /*! session id                   */
-	    is_exc_sess:1,        /*! if exception session         */
-	    error:1,              /*! error indication             */
+	u8  error:1,              /*! error indication             */
 	    drop:1;               /*! drop packet                  */
-	u32 hash_sig;             /*! hash signature result        */
-	u32 hash_h1;              /*! hash H1 result               */
-	u32 hash_h2;              /*! hash H2 result               */
-	u8  l3_off[PP_DESC_NUM_OF_L3_OFF];/*! parser L3 offsets    */
+	u8  l3_off[PP_DESC_NUM_OF_L3_OFF]; /*! parser L3 offsets   */
 	u8  tunn_off_id:3,        /*! L3 off id after tunnel hdr   */
 	    payld_off_id:3;       /*! L3 off id of payload         */
+#endif
 };
 
 /**
@@ -199,19 +202,21 @@ struct pp_pkt_ud {
 struct pp_desc {
 	u64 ps:48,           /*! protocol specific, up to 48bits         */
 	    data_off:9,      /*! data_pointer = buff_ptr + data_off      */
-	    pmac:1,          /*! 16B of PMAC header Part of Pre L2       */
-	    color:2,         /*! color at egress                         */
 	    src_pool:4;      /*! from which pool the buffer was taken    */
-	u64 buff_ptr:36,     /*! buffer pointer                          */
-	    pre_l2_sz:6,     /*! preL2 size incl PMAC hdr                */
-	    ts:1;            /*! indicate timestemp existence            */
+	u64 buff_ptr:36;     /*! buffer pointer                          */
 	u8  tx_port;         /*! Egress Port @ RXDMA output              */
 	u8  bm_policy;       /*! from which policy the buffer was taken  */
 	u16 pkt_len:14,      /*! packet length. Defaults incl preL2 & ts */
-	    lsp_pkt:1,       /*! last slow path packet, used for syncq   */
-	    own:1;           /*! descriptor ownership                    */
+	    lsp_pkt:1;       /*! last slow path packet, used for syncq   */
 	struct pp_pkt_ud ud; /*! packet user define block                */
 	u64 prv_ps:48;       /*! previous protocol specific (1st round)  */
+#ifdef PP_FULL_DESC_DBG
+	u8  pmac:1,           /*! 16B of PMAC header Part of Pre L2       */
+	    color:2,          /*! color at egress                         */
+	u8  pre_l2_sz:6,      /*! preL2 size incl PMAC hdr                */
+	    ts:1,             /*! indicate timestemp existence            */
+	    own:1;            /*! descriptor ownership                    */
+#endif
 };
 
 /**
@@ -334,6 +339,9 @@ struct pp_port_tx_info {
 
 	/*! if set this port can receive prel2 data */
 	bool prel2_en;
+
+	/*! if set, the RXDMA will copy the descriptor to the buffer */
+	bool wr_desc;
 };
 
 /**
@@ -592,6 +600,30 @@ struct pp_mcast_grp_info {
 #define PP_SESS_FLAG_LRO_INFO_MSK \
 	BIT(PP_SESS_FLAG_LRO_INFO_BIT)
 
+/**
+ * @define Specify if Turbodox prioritization should be used for this session
+ * @note dst_q_high must be configured
+ */
+#define PP_SESS_FLAG_TDOX_BIT                   (8)
+#define PP_SESS_FLAG_TDOX_MSK \
+	BIT(PP_SESS_FLAG_TDOX_BIT)
+
+/**
+ * @define Specify if Turbodox ack suppression should be used for this session
+ * @note dst_q_high must be configured
+ */
+#define PP_SESS_FLAG_TDOX_SUPP_BIT              (9)
+#define PP_SESS_FLAG_TDOX_SUPP_MSK \
+	BIT(PP_SESS_FLAG_TDOX_SUPP_BIT)
+
+/**
+ * @define Specify if hash will be calculated internally
+ * @note Step B Only!
+ */
+#define PP_SESS_FLAG_INTERNAL_HASH_CALC_BIT     (10)
+#define PP_SESS_FLAG_INTERNAL_HASH_CALC_MSK \
+	BIT(PP_SESS_FLAG_INTERNAL_HASH_CALC_BIT)
+
 /**
  * @struct pp_session_args
  * @brief PP session arguments required for creating a new
@@ -621,6 +653,11 @@ struct pp_sess_create_args {
 	/*! session's packets should be accelerated to */
 	u32 dst_q;
 
+	/*! logical high priority destination queue number where the */
+	/*! high priority session's packets should be accelerated to */
+	/*! relevant only for Turbodox sessions */
+	u32 dst_q_high;
+
 	/*! session group counters indexes, specify the group counters */
 	/*! to update for the session, use PP_SGC_INVALID to specify    */
 	/*! an invalid entry */
@@ -1270,4 +1307,4 @@ s32 pp_session_dst_queue_update(u32 id, u16 dst_q);
  * @return s32 on on success, error code otherwise
  */
 s32 pp_hw_clock_freq_set(u32 clk_MHz);
-#endif /* __PP_API_H__ */
+#endif /* __PP_API_H__ */
\ No newline at end of file
