From ad53d29b27526ac4fb11c08c798925de84df6604 Mon Sep 17 00:00:00 2001
From: agaraix <anath.bandux.garai@intel.com>
Date: Tue, 31 Mar 2020 19:13:18 +0530
Subject: [PATCH] LMTSYS-1754-ppv4-hal-ipsec-outbound-session-acceleration

---
 include/uapi/net/ppa_api.h | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index 8d2083c925f7..70622050f3aa 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -672,13 +672,8 @@ typedef struct {
 	uint64_t host_bytes; /*!< Number of bytes transmitted through CPU path */
 	uint64_t acc_bytes; /*!< Number of bytes accelerated by hardware engines */
 } PPA_SESSION_STATS;
-/*!
-	\brief This is the data structure for basic IPV4/IPV6 address
- */
-typedef union {
-	uint32_t ip; /*!< ipv4 address */
-	uint32_t ip6[4]; /*!< ipv6 address */
-} IP_ADDR;
+
+typedef PPA_IPADDR IP_ADDR;
 /*!
 	\brief This is the data structure for complex IPV4/IPV6 address
  */
