From c19ad115c550dd305bea846da4ca21716a4cbbbc Mon Sep 17 00:00:00 2001
From: vrp <vijaykumar.rp@intel.com>
Date: Thu, 16 Apr 2020 13:49:15 +0530
Subject: [PATCH] UGW_SW-48480 : Add Macro to specify SP+WRR scenario

---
 include/uapi/net/ppa_api.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index 35b3a2eb668d..caf7d5f3f104 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -417,6 +417,9 @@
 /*!
   \brief PPA_QOS_CL_CPU */
 #define PPA_QOS_CL_CPU				0x00010000
+/*!
+  \brief PPA_QOS_Q_F_SP_WFQ */
+#define PPA_QOS_Q_F_SP_WFQ	0x00020000 /*!< QoS configured in SP+WFQ mode */
 /*!
 	\brief PPA_F_PPPOATM */
 #define PPA_F_PPPOATM				0x80000000
