From 58492208dcde8d8017d7a8d1d644fd478e31870f Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Tue, 9 Jun 2020 18:31:53 +0800
Subject: [PATCH] DRVLIB_SW-3672: Use Macro for cpu vinit and veic

---
 arch/mips/lantiq/grx500/irq.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/arch/mips/lantiq/grx500/irq.c b/arch/mips/lantiq/grx500/irq.c
index cf3f9a9ac4fd..99a512eca705 100755
--- a/arch/mips/lantiq/grx500/irq.c
+++ b/arch/mips/lantiq/grx500/irq.c
@@ -19,12 +19,15 @@
 #include <asm/irq_cpu.h>
 #include <asm/mips-cm.h>
 
+#define CPU_VEIC_STATUS	(cpu_has_veic ? "on" : "off")
+#define CPU_VINT_STATUS	(cpu_has_vint ? "on" : "off")
+
 void __init arch_init_irq(void)
 {
 	struct device_node *intc_node;
 
-	pr_info("EIC is %s\n", cpu_has_veic ? "on" : "off");
-	pr_info("VINT is %s\n", cpu_has_vint ? "on" : "off");
+	pr_info("EIC is %s\n", CPU_VEIC_STATUS);
+	pr_info("VINT is %s\n", CPU_VINT_STATUS);
 
 	intc_node = of_find_compatible_node(NULL, NULL,
 					    "mti,cpu-interrupt-controller");
