From fc9853f108ee0bd55e4c267efe6697aba21a203b Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Wed, 10 Jun 2020 20:01:08 +0800
Subject: [PATCH] DRVLIB_SW-3672: Remove unreachable code in lantiq-vmb.c file

---
 arch/mips/lantiq/lantiq-vmb.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/arch/mips/lantiq/lantiq-vmb.c b/arch/mips/lantiq/lantiq-vmb.c
index 66e0c17a75f7..6c13b7d31c45 100644
--- a/arch/mips/lantiq/lantiq-vmb.c
+++ b/arch/mips/lantiq/lantiq-vmb.c
@@ -348,11 +348,9 @@ static int update_DB_from_DT(struct VMB_vpe_t *vt)
 	/* get the interrupt numbers from DT */
 	g_fw_vmb_irq[vt->cpu_id] = irqres.start;
 
-	ret = of_property_read_u32_index(np, "interrupts", 1,
-					 &g_fw_vmb_hwirq[vt->cpu_id]);
-
-	if (ret < 0) {
-		pr_info("failed to get hwirq for vmb since ret = %d\n", ret);
+	if (of_property_read_u32_index(np, "interrupts", 1,
+				       &g_fw_vmb_hwirq[vt->cpu_id]) != 0) {
+		pr_info("failed to get hwirq for vmb\n");
 		return -ENODEV;
 	}
 
