From 548b71a564b946c9d2f01989eec90f524ba769d0 Mon Sep 17 00:00:00 2001
From: yixin zhu <yixin.zhu@intel.com>
Date: Mon, 22 Jun 2020 13:25:09 +0800
Subject: [PATCH] poll dma controller reset status to confirm reset done

---
 drivers/dma/intel/hdma.c | 37 +++++++++++++++++++++++++------------
 1 file changed, 25 insertions(+), 12 deletions(-)

diff --git a/drivers/dma/intel/hdma.c b/drivers/dma/intel/hdma.c
index 725970431c88..044a4a8c52d7 100755
--- a/drivers/dma/intel/hdma.c
+++ b/drivers/dma/intel/hdma.c
@@ -21,6 +21,7 @@
 #include <linux/of.h>
 #include <linux/of_irq.h>
 #include <linux/slab.h>
+#include <linux/iopoll.h>
 
 #include "hdma.h"
 
@@ -403,11 +404,23 @@ static inline bool dma_chan_is_hw_desc(struct dmax_chan *pch)
 
 static void dma_ctrl_reset(struct dma_ctrl *pctrl)
 {
-	unsigned long flags;
+	unsigned long flags, val = 1;
+	int ret;
 
 	spin_lock_irqsave(&pctrl->ctrl_lock, flags);
 	ltq_dma_w32_mask(pctrl, 0, DMA_CTRL_RST, DMA_CTRL);
 	spin_unlock_irqrestore(&pctrl->ctrl_lock, flags);
+
+	/*
+	 * Delay maximum 10ms after reset to allow DMA logic to quiesce.
+	 * Do a polled read up to 1ms and 10ms maximum.
+	 */
+	ret = readl_poll_timeout(pctrl->membase + DMA_CTRL, val,
+				!(val & DMA_CTRL_RST), 1000, 10000);
+
+	if (ret)
+		dev_err(pctrl->dev, "%s controller reset timeout!\n",
+			pctrl->name);
 }
 
 /* DMA controller related configuration */
@@ -573,26 +586,32 @@ static int dma_ctrl_cfg(struct dma_ctrl *pctrl)
 {
 	int enable;
 
-	if ((pctrl->flags & DMA_FLCTL))
+	if (pctrl->flags & DMA_FLCTL)
 		enable = 1;
 	else
 		enable = 0;
 	dma_ctrl_chan_flow_ctl_cfg(pctrl, enable);
 
-	if ((pctrl->flags & DMA_DESC_IN_SRAM))
+	if (pctrl->flags & DMA_FTOD)
+		enable = 1;
+	else
+		enable = 0;
+	dma_ctrl_desc_fetch_on_demand_cfg(pctrl, enable);
+
+	if (pctrl->flags & DMA_DESC_IN_SRAM)
 		enable = 1;
 	else
 		enable = 0;
 	dma_ctrl_sram_desc_cfg(pctrl, enable);
 
 	dma_ctrl_arb_cfg(pctrl);
-	if ((pctrl->flags & DMA_DRB))
+	if (pctrl->flags & DMA_DRB)
 		enable = 1;
 	else
 		enable = 0;
 	dma_ctrl_desc_read_back_cfg(pctrl, enable);
 
-	if ((pctrl->flags & DMA_EN_BYTE_EN))
+	if (pctrl->flags & DMA_EN_BYTE_EN)
 		enable = 1;
 	else
 		enable = 0;
@@ -2983,7 +3002,7 @@ static irqreturn_t dma_chan_interrupt(int irq, void *dev_id)
 static int dma_ctrl_init(struct dma_ctrl *pctrl)
 {
 	u32 i, j;
-	int enable, ret;
+	int ret;
 	struct dma_port *pport = NULL;
 	struct dmax_chan *pch = NULL;
 
@@ -3019,12 +3038,6 @@ static int dma_ctrl_init(struct dma_ctrl *pctrl)
 		}
 	}
 
-	if ((pctrl->flags & DMA_FTOD))
-		enable = 1;
-	else
-		enable = 0;
-	dma_ctrl_desc_fetch_on_demand_cfg(pctrl, enable);
-
 	return 0;
 }
 
