From 3aa617d1b4c61f98062c0b2725fa7eb27ec12942 Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Mon, 20 Jul 2020 18:01:12 +0200
Subject: [PATCH] PONRTSYS-7614: xpcs: Ignore tries to set speed 0

---
 drivers/net/ethernet/lantiq/xpcs/xpcs.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
index fdfada0ad2a6..e61071b6fcf2 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
@@ -1553,6 +1553,12 @@ static int xpcs_ethtool_ksettings_set(struct device *dev,
 		return -1;
 	}
 
+	/* Ignore settings with speed = 0, which can happen if settings were get
+	 * before a valid interface speed was detected and then applied back.
+	 */
+	if (speed == 0)
+		return 0;
+
 	if (speed != SPEED_10000 &&
 	    speed != SPEED_1000 &&
 	    speed != SPEED_100 &&
