From ae3c8d6f7486fa62e22f342f128ee285cd30e52b Mon Sep 17 00:00:00 2001
From: "Shao, Guohua" <gshao@maxlinear.com>
Date: Mon, 28 Jun 2021 17:02:18 +0800
Subject: [PATCH] DRV_DPM_SW-519 dpm add new flag for IP host

---
 include/net/datapath_api.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index 7b8ae6fb7ec9..86faade97d55 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -865,6 +865,10 @@ enum DP_SUBIF_DATA_FLAG {
 	DP_SUBIF_SEG_DIS = BIT(12), /*!< if this flag is set seg_en=0 to use
 								 * BM buffer
 								 */
+	DP_SUBIF_NON_CPU_MAC = BIT(13), /*!< if this flag is set, dpm will not
+					 *   set this device's mac address to
+					 *   CPU bridge port
+					 */
 };
 
 /*! @brief dp_subif_id struct for get_netif_subif */
