# HG changeset patch
# Parent  f0e81211258f1a333174c26fe79dd9a4ade7894a

diff --git a/drivers/timer/tsc_timer.c b/drivers/timer/tsc_timer.c
--- a/drivers/timer/tsc_timer.c
+++ b/drivers/timer/tsc_timer.c
@@ -404,6 +404,11 @@ static void tsc_timer_ensure_setup(bool 
 	if (!gd->arch.clock_rate) {
 		unsigned long fast_calibrate;
 
+		if (early) {
+			gd->arch.clock_rate = CONFIG_X86_TSC_TIMER_FREQ;
+			goto done;
+		}
+
 		fast_calibrate = native_calibrate_tsc();
 		if (fast_calibrate)
 			goto done;
@@ -424,15 +429,11 @@ static void tsc_timer_ensure_setup(bool 
 		if (fast_calibrate)
 			goto done;
 
-		if (early)
-			gd->arch.clock_rate = CONFIG_X86_TSC_TIMER_FREQ;
-		else
-			return;
-
 done:
 		if (!gd->arch.clock_rate)
 			gd->arch.clock_rate = fast_calibrate * 1000000;
 	}
+
 	gd->arch.tsc_inited = true;
 }
 
@@ -449,6 +450,10 @@ static int tsc_timer_probe(struct udevic
 		 */
 		if (!uc_priv->clock_rate)
 			panic("TSC frequency is ZERO");
+
+		/* set to the correct tsc timer value from dts when on HAPS */
+		if (CONFIG_IS_ENABLED(IS_HAPS))
+			gd->arch.clock_rate = uc_priv->clock_rate;
 	} else {
 		uc_priv->clock_rate = gd->arch.clock_rate;
 	}
