# HG changeset patch
# Parent  93d92c5e07b3d46d590f10069261d9232336f077

diff --git a/cmd/mxl/cmd_icc.c b/cmd/mxl/cmd_icc.c
new file mode 100644
--- /dev/null
+++ b/cmd/mxl/cmd_icc.c
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright © 2022 MaxLinear, Inc.
+ *
+ */
+#include <common.h>
+#include <command.h>
+
+extern int icc_uboot(int icc_type, u32 icc_data0, u32 icc_data1, u32 icc_data2, u32 icc_data3);
+
+int do_icc(struct cmd_tbl *cmdtp, int flag, int argc, char * const argv[])
+{
+	int type, ret;
+	u32 data0, data1, data2, data3;
+
+	if (argc != 6)
+		return CMD_RET_USAGE;
+
+	type  = simple_strtoul(argv[1], NULL, 16);
+	data0 = simple_strtoul(argv[2], NULL, 16);
+	data1 = simple_strtoul(argv[3], NULL, 16);
+	data2 = simple_strtoul(argv[4], NULL, 16);
+	data3 = simple_strtoul(argv[5], NULL, 16);
+
+	ret = icc_uboot(type, data0, data1, data2, data3);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+U_BOOT_CMD(
+	icc, CONFIG_SYS_MAXARGS, 0, do_icc,
+	"send icc message with boot informations to tep",
+	"<message type> <data 0> <data 1> <data 2> <data 3>\n"
+	"\t- message type (SYNC_TEP=1, UPGRADE=2, CFG_FIREWALL=3)\n"
+	"\t- <SYNC_TEP>     <kernel start addr> <kernel end addr> <rootfs start addr> <max memory size in Mbytes>\n"
+	"\t- <UPGRADE>      <upgrade time (secs)> <null> <null> <null>\n"
+	"\t- <CFG_FIREWALL> <kernel start addr> <kernel end addr> <null> <null>\n"
+);
