# HG changeset patch
# Parent  1fb7cc25e766a9df719d855a9ad399143f4b5341

diff --git a/drivers/mtd/nand/spi/mxl_spi_nand.c b/drivers/mtd/nand/spi/mxl_spi_nand.c
new file mode 100644
--- /dev/null
+++ b/drivers/mtd/nand/spi/mxl_spi_nand.c
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2022 Maxlinear Corporation
+ *
+ */
+
+#include <common.h>
+#include <mtd.h>
+#include <linux/errno.h>
+#include <asm/arch/project.h>
+
+int board_nand_init(void)
+{
+	struct mtd_info *mtd_data;
+
+	/* probe all MTD devices */
+	mtd_probe_devices();
+
+	mtd_data = get_mtd_device_nm("spi-nand0");
+
+	/* Device */
+	rbe_debug_print("\n SPL: Device: %s\n", mtd_data->dev->name);
+
+	/* MTD device information */
+	rbe_debug_print(" SPL: Type: ");
+	switch (mtd_data->type) {
+	case MTD_NORFLASH:
+		rbe_debug_print("NOR flash\n");
+		break;
+	case MTD_NANDFLASH:
+		rbe_debug_print("NAND flash\n");
+		break;
+	case MTD_ABSENT:
+	default:
+		rbe_debug_print("Unknown\n");
+		break;
+	}
+
+	rbe_debug_print(" SPL: Block size: 0x%x bytes\n", mtd_data->erasesize);
+	rbe_debug_print(" SPL: Min I/O: 0x%x bytes\n", mtd_data->writesize);
+	rbe_debug_print(" SPL: OOB size: %u bytes\n", mtd_data->oobsize);
+	return 0;
+}
