From 47d5deb466357825b8a1c170601f4a51de639075 Mon Sep 17 00:00:00 2001
From: Dmitrijs Martinovs <dmartinovs@maxlinear.com>
Date: Thu, 8 Jun 2023 14:09:42 +0200
Subject: [PATCH] WLANRTSYS-56024: Map MXL proprietary configuration to
 opensource part 2

---
 hostapd/config_file.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index e399d6d56..0749f8d03 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5392,7 +5392,8 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		if (get_hex_config(bss->ext_capa, EXT_CAPA_MAX_LEN,
 				   line, "ext_capa", pos))
 			return 1;
-	} else if (os_strcmp(buf, "rnr") == 0) {
+	// TODO: tmp solution - remake this to work with 1 parameter name
+	} else if (os_strcmp(buf, "rnr") == 0 || os_strcmp(buf, "rnr_auto_update") == 0) {
 		bss->rnr = atoi(pos);
 #ifdef CONFIG_IEEE80211BE
 	} else if (os_strcmp(buf, "ieee80211be") == 0) {
-- 
2.43.0

