From 1b54c4ca8e48a0c55bf8c9a4e13aae43b064ff56 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 9 Jun 2023 09:18:16 +0200
Subject: [PATCH] WLANRTSYS-55992: Implement radius greylist access control:
 part1

---
 hostapd/config_file.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 7929a2e3d..b0c2472d9 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5550,6 +5550,16 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 			return 1;
 		}
 		bss->opmode_notif = val;
+	} else if (os_strcmp(buf, "greylist") == 0) {
+			wpa_printf(MSG_WARNING, "Line %d: greylist: greylist feature is not implemented yet", line);
+	} else if (os_strcmp(buf, "ap_vlan") == 0) {
+			wpa_printf(MSG_WARNING, "Line %d: ap_vlan: greylist feature is not implemented yet", line);
+	} else if (os_strcmp(buf, "cmmac") == 0) {
+			wpa_printf(MSG_WARNING, "Line %d: cmmac: greylist feature is not implemented yet", line);
+	} else if (os_strcmp(buf, "greylist_logpath") == 0) {
+			wpa_printf(MSG_WARNING, "Line %d: greylist_logpath: greylist feature is not implemented yet", line);
+	} else if (os_strcmp(buf, "greylist_file") == 0) {
+			wpa_printf(MSG_WARNING, "Line %d: greylist_file: greylist feature is not implemented yet", line);
 #ifdef CONFIG_VENDOR_MXL
 	} else if (mxl_hostapd_config_fill(conf, bss, buf, pos, line)) {
 #endif /* CONFIG_VENDOR_MXL */
-- 
2.43.0

