From f8404fdbc115de12c273e412d585b77824a50207 Mon Sep 17 00:00:00 2001
From: Dmitrijs Martinovs <dmartinovs@maxlinear.com>
Date: Tue, 20 Jun 2023 13:52:09 +0200
Subject: [PATCH] WLANRTSYS-56281: Add rssi_ignore_probe_request pipe
 functionality

---
 src/ap/hostapd.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 396547d60..8576d46e9 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1633,6 +1633,13 @@ static int hostapd_setup_bss(struct hostapd_data *hapd, int first,
 		return -1;
 #endif /* CONFIG_VENDOR_MXL */
 
+	if (hapd->iconf->rssi_ignore_probe_request != 0){
+		if (hostapd_drv_vendor_cmd(hapd, OUI_LTQ, LTQ_NL80211_VENDOR_SUBCMD_SET_RSSI_IGNORE_PROBE_REQ,
+				(u8*)&hapd->iconf->rssi_ignore_probe_request,
+				sizeof(hapd->iconf->rssi_ignore_probe_request), NESTED_ATTR_NOT_USED, NULL))
+			wpa_printf(MSG_WARNING, "set rssi_ignore_probe_request failed");
+	}
+
 	if (conf->wmm_enabled < 0)
 		conf->wmm_enabled = hapd->iconf->ieee80211n |
 			hapd->iconf->ieee80211ax;
-- 
2.43.0

