From 117c699cb1321ef36bba8d04da47a5bcb215b0f6 Mon Sep 17 00:00:00 2001
From: mvulsx <mvuls@maxlinear.com>
Date: Fri, 30 Jun 2023 05:34:27 +0000
Subject: [PATCH] WLANRTSYS-56610: Reject AssocReq with zero length supported
 rates information element

Signed-off-by: mvulsx <mvuls@maxlinear.com>
---
 src/ap/ieee802_11.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index c8a2b7217..431de4624 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3520,6 +3520,13 @@ static u16 copy_supp_rates(struct hostapd_data *hapd, struct sta_info *sta,
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
 	}
 
+	if (elems->supp_rates_len == 0) {
+		hostapd_logger(hapd, sta->addr, HOSTAPD_MODULE_IEEE80211,
+			HOSTAPD_LEVEL_INFO,
+			"Zero length of supported rates element in AssocReq");
+		return WLAN_STATUS_INVALID_IE;
+	}
+
 	if (elems->supp_rates_len + elems->ext_supp_rates_len >
 	    sizeof(sta->supported_rates)) {
 		hostapd_logger(hapd, sta->addr, HOSTAPD_MODULE_IEEE80211,
-- 
2.43.0

