From 2e90f8def7e87939098fe5ff4e0f385b97a45194 Mon Sep 17 00:00:00 2001
From: hmgokhale <hmgokhale@maxlinear.com>
Date: Tue, 18 Jun 2024 14:49:10 +0530
Subject: [PATCH] WLANRTSYS-81439 Shared header changes

Signed-off-by: hmgokhale <hmgokhale@maxlinear.com>
---
 src/common/vendor_cmds_copy.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/common/vendor_cmds_copy.h b/src/common/vendor_cmds_copy.h
index 0350a94c1..31fd08aa3 100644
--- a/src/common/vendor_cmds_copy.h
+++ b/src/common/vendor_cmds_copy.h
@@ -408,6 +408,7 @@ enum ltq_nl80211_vendor_subcmds {
   LTQ_NL80211_VENDOR_SUBCMD_ML_SEND_T2LM_TEARDOWN_FRAME,
   LTQ_NL80211_VENDOR_SUBCMD_GET_ML_LINK_STATS,
   LTQ_NL80211_VENDOR_SUBCMD_SET_ML_LINK_STATS,
+  LTQ_NL80211_VENDOR_SUBCMD_GET_ML_VAP_LIST,
   LTQ_NL80211_VENDOR_SUBCMD_SET_DYNAMIC_EDCA,
   LTQ_NL80211_VENDOR_SUBCMD_SET_STR_TID_LINK_SPREADING,
   LTQ_NL80211_VENDOR_SUBCMD_REG_UPDATE_6GHZ_OPER_POWER_MODE,
-- 
2.43.0

