From 1a82562293517744904acc1606c4ee3643d64840 Mon Sep 17 00:00:00 2001
From: npandian <npandian@maxlinear.com>
Date: Wed, 31 Jul 2024 12:31:50 +0530
Subject: [PATCH] WLANRTSYS-60538 Aligning shared headers

---
 src/common/vendor_cmds_copy.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/common/vendor_cmds_copy.h b/src/common/vendor_cmds_copy.h
index 31fd08aa3..e0bfbcb42 100644
--- a/src/common/vendor_cmds_copy.h
+++ b/src/common/vendor_cmds_copy.h
@@ -418,6 +418,8 @@ enum ltq_nl80211_vendor_subcmds {
   LTQ_NL80211_VENDOR_SUBCMD_GET_ML_STA_LIST,
   LTQ_NL80211_VENDOR_SUBCMD_SET_RADAR_FIFO_DUMP,
   LTQ_NL80211_VENDOR_SUBCMD_SET_DUPLICATE_BEACON,
+  LTQ_NL80211_VENDOR_SUBCMD_GET_PCIE_AUTO_GEN_ENABLE,
+  LTQ_NL80211_VENDOR_SUBCMD_SET_PCIE_AUTO_GEN_ENABLE,
   /* add new iw sub commands here */
 
   /* used to define LTQ_NL80211_VENDOR_SUBCMD_MAX below */
-- 
2.43.0

