From 48e03a807e0391c7087415c68d3c0794087411fb Mon Sep 17 00:00:00 2001
From: Bharath Paidupati <bpaidupati@maxlinear.com>
Date: Tue, 8 Oct 2024 13:03:43 +0530
Subject: [PATCH] WLANRTSYS-85179 Adding 320Mhz support in afc.c file

---
 src/ap/afc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/afc.c b/src/ap/afc.c
index 6513d07ae..cb27e94bc 100644
--- a/src/ap/afc.c
+++ b/src/ap/afc.c
@@ -235,6 +235,11 @@ static struct json_object * hostapd_afc_get_opclass_chan_list(u8 op_class)
 	case 134: /* 160MHz */
 		chan_offset = 14;
 		break;
+#ifdef CONFIG_VENDOR_MXL
+	case 137: /* 320MHz */
+		chan_offset = 30;
+		break;
+#endif /* CONFIG_VENDOR_MXL */
 	default:
 		chan_offset = 0;
 		break;
-- 
2.43.0

