--- a/cache.c
+++ b/cache.c
@@ -386,6 +386,7 @@ cache_dump_records(struct blob_buf *buf, const char *name)
 {
 	struct cache_record *r, *last, *next;
 	const char *txt;
+	void *l;
 	char buffer[INET6_ADDRSTRLEN];
 
 	last = avl_last_element(&records, last, avl);
@@ -393,12 +394,13 @@ cache_dump_records(struct blob_buf *buf, const char *name)
 		const struct dns_srv_data *dsd = (const struct dns_srv_data*)r->rdata;
 		switch (r->type) {
 		case TYPE_TXT:
-			if (r->txt && strlen(r->txt)) {
-				txt = r->txt;
+			if ((txt = r->txt) && strlen(r->txt)) {
+				l = blobmsg_open_array(buf, "txt");
 				do {
-					blobmsg_add_string(buf, "txt", txt);
+					blobmsg_add_string(buf, NULL, txt);
 					txt = &txt[strlen(txt) + 1];
 				} while (*txt);
+				blobmsg_close_array(buf, l);
 			}
 			break;
 