From c9e9c75de20dbd633bec90cf3ccfc6a36af39ff4 Mon Sep 17 00:00:00 2001
From: "Rywacki, Kamil" <krywacki@maxlinear.com>
Date: Fri, 29 Jul 2022 11:15:00 +0200
Subject: bridge: Port linux kernel v5.15 flags

---
 include/linux-private/linux/if_link.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/linux-private/linux/if_link.h b/include/linux-private/linux/if_link.h
index 7f0eae7..e5084fb 100644
--- a/include/linux-private/linux/if_link.h
+++ b/include/linux-private/linux/if_link.h
@@ -347,6 +347,10 @@ enum {
 	IFLA_BRPORT_NEIGH_SUPPRESS,
 	IFLA_BRPORT_ISOLATED,
 	IFLA_BRPORT_BACKUP_PORT,
+	IFLA_BRPORT_MRP_RING_OPEN,
+	IFLA_BRPORT_MRP_IN_OPEN,
+	IFLA_BRPORT_MCAST_EHT_HOSTS_LIMIT,
+	IFLA_BRPORT_MCAST_EHT_HOSTS_CNT,
 	IFLA_BRPORT_LEARNING_LIMIT,
 	IFLA_BRPORT_DEST_LOOKUP,
 	IFLA_BRPORT_SRC_LOOKUP,
