From c89428ffa50b93c6e3bc2a6f47af2ab2a14599c9 Mon Sep 17 00:00:00 2001
From: Radhakrishna Ravipati <radha.ravipatix.krishna@intel.com>
Date: Thu, 9 Sep 2021 12:16:01 +0530
Subject: Disable WEXT_PROC

---
 net/wireless/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/wireless/Kconfig b/net/wireless/Kconfig
index 3e14a14ef..ceeb8daa7 100644
--- a/net/wireless/Kconfig
+++ b/net/wireless/Kconfig
@@ -7,7 +7,6 @@ config WEXT_CORE
 	depends on CFG80211_WEXT || WIRELESS_EXT
 
 config WEXT_PROC
-	def_bool y
 	depends on PROC_FS
 	depends on WEXT_CORE
 
-- 
2.17.1

