--- a/src/linux/ifxos_linux_copy_user_space_drv.c	2011-03-19 13:34:33.000000000 +0530
+++ b/src/linux/ifxos_linux_copy_user_space_drv.c	2018-12-27 20:44:20.000000000 +0530
@@ -29,7 +29,12 @@
 #ifdef MODULE
    #include <linux/module.h>
 #endif
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 12, 0)
 #include <asm/uaccess.h>
+#else
+#include <linux/uaccess.h>
+#endif
 
 #include "ifx_types.h"
 #include "ifxos_rt_if_check.h"
