From 954ae8b80c1cc4a5c21b61145e1a470a4dbf8266 Mon Sep 17 00:00:00 2001
From: Hua Ma <hua.ma@linux.intel.com>
Date: Thu, 21 Jun 2018 17:38:08 +0800
Subject: [PATCH] Add support for spi common files

---
 drivers/spi/Kconfig  | 44 ++++++++++++++++++++++++++++++++++++++++++++
 drivers/spi/Makefile |  4 ++++
 2 files changed, 48 insertions(+)

diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index e9576658ff3f..391af1fdc8dc 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -192,6 +192,12 @@ config SPI_CADENCE
 	  This selects the Cadence SPI controller master driver
 	  used by Xilinx Zynq and ZynqMP.
 
+config SPI_CADENCE_QSPI
+	tristate "Cadence QSPI controller"
+	help
+	  This enables support for the Cadence QSPI controller in master
+	  mode.
+
 config SPI_CLPS711X
 	tristate "CLPS711X host SPI controller"
 	depends on ARCH_CLPS711X || COMPILE_TEST
@@ -280,6 +286,20 @@ config SPI_GPIO
 	  GPIO operations, you should be able to leverage that for better
 	  speed with a custom version of this driver; see the source code.
 
+config SPI_GRX500
+	tristate "Lantiq GRX500 SPI controller"
+	select SPI_BITBANG
+	help
+	  This driver supports the Lantiq GRX500 SoC SPI controller in master mode.
+
+config SPI_GRX500_POLL
+	bool "Lantiq SPI/SSC polling mode"
+	depends on SPI_GRX500
+	default n
+	help
+	  Selecting this will do the read and write in polling mode or
+	  manually checking the FIFO level until avail.
+
 config SPI_IMG_SPFI
 	tristate "IMG SPFI controller"
 	depends on MIPS || COMPILE_TEST
@@ -563,6 +583,30 @@ config SPI_SC18IS602
 	help
 	  SPI driver for NXP SC18IS602/602B/603 I2C to SPI bridge.
 
+config LANTIQ_SPI
+	tristate "Lantiq SPI/SSC driver"
+	depends on SPI_MASTER
+	default n
+
+config LANTIQ_SPI_DEBUG
+    bool "Lantiq SPI/SSC runtime debug support"
+    depends on LANTIQ_SPI
+    default n
+    help
+    Choose this option to enable Lantiq SPI/SSC runtime debug options.
+    This can be enabled by writing debug options into
+    /proc/sys/dev/ltq_ssc/debug.
+
+config LANTIQ_SPI_ASYNCHRONOUS
+	bool "Lantiq SPI/SSC tasklet support"
+	depends on LANTIQ_SPI
+	default y
+	---help---
+	Choose this option to support synchronous and asynchronous APIs
+	at the same time. SPI itself will run in tasklet. If not,
+	only synchronous API is supported. SPI context will in kernel
+	thread mode and belong to caller's context.
+
 config SPI_SH_MSIOF
 	tristate "SuperH MSIOF SPI controller"
 	depends on HAVE_CLK && HAS_DMA
diff --git a/drivers/spi/Makefile b/drivers/spi/Makefile
index aa939d955521..91925b8511a8 100644
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -45,6 +45,7 @@ obj-$(CONFIG_SPI_FSL_LIB)		+= spi-fsl-lib.o
 obj-$(CONFIG_SPI_FSL_ESPI)		+= spi-fsl-espi.o
 obj-$(CONFIG_SPI_FSL_SPI)		+= spi-fsl-spi.o
 obj-$(CONFIG_SPI_GPIO)			+= spi-gpio.o
+obj-$(CONFIG_SPI_GRX500)        += spi-grx500.o
 obj-$(CONFIG_SPI_IMG_SPFI)		+= spi-img-spfi.o
 obj-$(CONFIG_SPI_IMX)			+= spi-imx.o
 obj-$(CONFIG_SPI_JCORE)			+= spi-jcore.o
@@ -102,3 +103,6 @@ obj-$(CONFIG_SPI_XILINX)		+= spi-xilinx.o
 obj-$(CONFIG_SPI_XLP)			+= spi-xlp.o
 obj-$(CONFIG_SPI_XTENSA_XTFPGA)		+= spi-xtensa-xtfpga.o
 obj-$(CONFIG_SPI_ZYNQMP_GQSPI)		+= spi-zynqmp-gqspi.o
+obj-$(CONFIG_SPI_CADENCE_QSPI)		+= spi-cadence-qspi.o
+obj-$(CONFIG_SPI_CADENCE_QSPI)		+= spi-cadence-qspi-apb.o
+obj-$(CONFIG_LANTIQ_SPI)		+= ltq_ssc.o
