From 376717aa4200fe808c4bc045c68ceb7352678e64 Mon Sep 17 00:00:00 2001
From: Hua Ma <hua.ma@linux.intel.com>
Date: Thu, 21 Jun 2018 17:38:16 +0800
Subject: [PATCH] Add support for crypto common files

---
 drivers/crypto/Kconfig  | 42 ++++++++++++++++++++++++++++++++++++++++++
 drivers/crypto/Makefile |  2 ++
 2 files changed, 44 insertions(+)

diff --git a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
index 4d2b81f2b223..cd7958c1c149 100644
--- a/drivers/crypto/Kconfig
+++ b/drivers/crypto/Kconfig
@@ -555,4 +555,46 @@ config CRYPTO_DEV_ROCKCHIP
 
 source "drivers/crypto/chelsio/Kconfig"
 
+config CRYPTO_DEV_LANTIQ_EIP123
+   tristate "Support Lantiq EIP123 hardware crypto engine"
+   depends on LANTIQ && SOC_TYPE_GRX500_TEP
+   default n
+   help
+      Selects Hardware accelerated EIP123 crypto drivers.
+
+config CRYPTO_DEV_LANTIQ_EIP97
+   tristate "Support Lantiq EIP97 hardware crypto engine"
+   depends on LANTIQ && SOC_GRX500
+   select CRYPTO_AUTHENC
+   select CRYPTO_SHA1
+   select CRYPTO_SHA256
+   select CRYPTO_SHA512
+   select CRYPTO_MD5
+   select CRYPTO_AES
+   select CRYPTO_DES
+   select CRYPTO_MANAGER_DISABLE_TESTS
+   default n
+   help
+      Selects Hardware accelerated EIP97 crypto drivers.
+
+config LTQ_CRYPTO_TEST
+   tristate "Lantiq Crypto Test"
+   depends on m && CRYPTO_DEV_LANTIQ_EIP97
+   help
+      Test suites for the hw crypto algs
+
+config LTQ_MPE_IPSEC_SUPPORT
+   bool
+   depends on LTQ_PPA_MPE_IP97
+   default y
+
+config LTQ_CRYPTO_MAX_RING_USED
+   int "Maximum number of ring used in the driver"
+   depends on CRYPTO_DEV_LANTIQ_EIP97
+   default "4" if LTQ_MPE_IPSEC_SUPPORT
+   default "2"
+   help
+     Number of rings used in the driver. By default, the driver supports up to
+     two rings. However, if MPE firmware is used, we only use 1 ring in the driver
+
 endif # CRYPTO_HW
diff --git a/drivers/crypto/Makefile b/drivers/crypto/Makefile
index ad7250fa1348..f78abb7bf059 100644
--- a/drivers/crypto/Makefile
+++ b/drivers/crypto/Makefile
@@ -32,3 +32,5 @@ obj-$(CONFIG_CRYPTO_DEV_VMX) += vmx/
 obj-$(CONFIG_CRYPTO_DEV_SUN4I_SS) += sunxi-ss/
 obj-$(CONFIG_CRYPTO_DEV_ROCKCHIP) += rockchip/
 obj-$(CONFIG_CRYPTO_DEV_CHELSIO) += chelsio/
+obj-$(CONFIG_CRYPTO_DEV_LANTIQ_EIP123) += lantiq_eip123/
+obj-$(CONFIG_CRYPTO_DEV_LANTIQ_EIP97) += lantiq_eip97/
