From 80a22df3b12d788f96cf250c833a3d3d340fec3f Mon Sep 17 00:00:00 2001
From: Hua Ma <hua.ma@linux.intel.com>
Date: Thu, 21 Jun 2018 17:38:19 +0800
Subject: [PATCH] Add support for arch x86 cmommon files

---
 arch/x86/kernel/tsc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kernel/tsc.c b/arch/x86/kernel/tsc.c
old mode 100644
new mode 100755
index cb768417429d..61f4500a8bea
--- a/arch/x86/kernel/tsc.c
+++ b/arch/x86/kernel/tsc.c
@@ -329,7 +329,7 @@ unsigned long long sched_clock(void)
 }
 #else
 unsigned long long
-sched_clock(void) __attribute__((alias("native_sched_clock")));
+sched_clock(void) __attribute__((weak, alias("native_sched_clock")));
 #endif
 
 int check_tsc_unstable(void)
