From 60319b3d66bbf9c44508a3c5b1584065c348f47b Mon Sep 17 00:00:00 2001
From: zhu yixin <yixin.zhu@intel.com>
Date: Tue, 3 Jul 2018 17:55:33 +0800
Subject: [PATCH] fix wait pll lock timeout issue.

---
 drivers/clk/intel/clk_api.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/clk/intel/clk_api.c b/drivers/clk/intel/clk_api.c
index 10de6538c91b..4d7b82ac9c7c 100644
--- a/drivers/clk/intel/clk_api.c
+++ b/drivers/clk/intel/clk_api.c
@@ -1065,7 +1065,7 @@ pll_clk_set(struct regmap *map, u32 reg,
 	    spinlock_t *lock)
 {
 	unsigned int mult, div, frac;
-	const int loop_cnt = 8;
+	const int loop_cnt = 100;
 	int i;
 
 	mult = ff_tbl->mult;
