From 3714d74272761add847353f90f4e5d59efb3d215 Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Wed, 1 Aug 2018 12:35:26 +0200
Subject: [PATCH] thermal: ltq: Fix checkpatch errors

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/thermal/ltq_thermal.c | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/drivers/thermal/ltq_thermal.c b/drivers/thermal/ltq_thermal.c
index 6b06470d6a15..2438926c6545 100644
--- a/drivers/thermal/ltq_thermal.c
+++ b/drivers/thermal/ltq_thermal.c
@@ -101,7 +101,7 @@ void  ltq_grx500_init(struct ltq_thermal_sensor *sensor)
 
 	/* Power on the sensor */
 	regmap_update_bits(priv->chiptop, CTRL_REG, TS_EN_MASK,
-				 TS_EN_WORKING);
+			   TS_EN_WORKING);
 	usleep_range(160, 300);  /* required by TS spec */
 
 	/* Start data conversion */
@@ -110,7 +110,7 @@ void  ltq_grx500_init(struct ltq_thermal_sensor *sensor)
 
 	/* Stop data conversion and power down the sensor*/
 	regmap_update_bits(priv->chiptop, CTRL_REG, SOC_MASK | TS_EN_MASK,
-			   SOC_NC|TS_EN_SHUNT);
+			   SOC_NC | TS_EN_SHUNT);
 }
 
 int ltq_grx500_get_temp(struct ltq_thermal_sensor *sensor)
@@ -131,9 +131,8 @@ int ltq_grx500_get_temp(struct ltq_thermal_sensor *sensor)
 			   sensor->id << CH_SEL_OFFSET);
 
 	/* Enable the temp sensor */
-	regmap_update_bits(priv->chiptop, CTRL_REG, SOC_MASK|TS_EN_MASK,
-				SOC_START | TS_EN_WORKING);
-
+	regmap_update_bits(priv->chiptop, CTRL_REG, SOC_MASK | TS_EN_MASK,
+			   SOC_START | TS_EN_WORKING);
 
 	/* Wait for and read out the measurement */
 	while (!ready) {
@@ -145,14 +144,14 @@ int ltq_grx500_get_temp(struct ltq_thermal_sensor *sensor)
 	v = reg & TS_CODE;
 
 	/* Disable the temp sensor */
-	regmap_update_bits(priv->chiptop, CTRL_REG, SOC_MASK|TS_EN_MASK,
-				SOC_NC|TS_EN_SHUNT);
+	regmap_update_bits(priv->chiptop, CTRL_REG, SOC_MASK | TS_EN_MASK,
+			   SOC_NC | TS_EN_SHUNT);
 
 	/* Temperature interpolation */
 	T = (int)(a0 * 1000000 + (-435356 * (v - v1)) +
 		  (-360 * (v - v1) * (v - v2)) + 1);
 
-	return T/1000;  /* returns °mC */
+	return T / 1000;  /* returns °mC */
 }
 
 /* Temperature sensor specific data */
@@ -211,6 +210,7 @@ static int ltq_thermal_get_trend(void *data, int trip,
 
 	return 0;
 }
+
 static int ltq_thermal_set_emul_temp(void *data, int temp)
 {
 	struct ltq_thermal_sensor *sensor = data;
@@ -239,7 +239,6 @@ static const struct of_device_id ltq_thermal_match[] = {
 
 static int ltq_thermal_probe(struct platform_device *pdev)
 {
-
 	struct ltq_thermal *priv;
 	struct device_node *node = pdev->dev.of_node;
 	const struct ltq_thermal_sensor_ops *sens_ops =
