From e3522141b7659eafe877bd6d6a97d80bf17eaa82 Mon Sep 17 00:00:00 2001
From: Wu Zhixian <zhixianx.wu@intel.com>
Date: Thu, 2 Aug 2018 19:40:16 +0800
Subject: [PATCH] fix crush issue

---
 drivers/dma/ltq_hwmcpy.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/dma/ltq_hwmcpy.c b/drivers/dma/ltq_hwmcpy.c
index 0244b399779e..7b586b71fa75 100644
--- a/drivers/dma/ltq_hwmcpy.c
+++ b/drivers/dma/ltq_hwmcpy.c
@@ -1348,7 +1348,7 @@ static int mcpy_xrx500_probe(struct platform_device *pdev)
 	/* Link platform with driver data for retrieving */
 	platform_set_drvdata(pdev, pctrl);
 	mcpy_proc_init(pctrl);
-	#ifdef CONFIG_LTQ_UMT_518_FW_SG
+	#if 0 // def CONFIG_LTQ_UMT_518_FW_SG
 	setup_tc_percpu((void *)pctrl, 0);
 	#else
 	umt_init(pctrl);
