From 9efb21eb61d7bb8636d8906c086097dacabc28f6 Mon Sep 17 00:00:00 2001
From: moinakde <moinak.debnath@intel.com>
Date: Fri, 3 Aug 2018 10:42:47 +0530
Subject: [PATCH] UGW:29102 adding Kconfig for eip123

---
 drivers/crypto/Kconfig | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
index 4d2b81f2b223..93df829868da 100644
--- a/drivers/crypto/Kconfig
+++ b/drivers/crypto/Kconfig
@@ -555,4 +555,11 @@ config CRYPTO_DEV_ROCKCHIP
 
 source "drivers/crypto/chelsio/Kconfig"
 
+config CRYPTO_DEV_LANTIQ_EIP123
+   tristate "Support Lantiq EIP123 hardware crypto engine"
+   depends on LANTIQ && SOC_TYPE_GRX500_TEP
+   default n
+   help
+      Selects Hardware accelerated EIP123 crypto drivers.
+
 endif # CRYPTO_HW
