From 099d9dee23248dbffa3d6773fbf8202dabcde81a Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Mon, 3 Sep 2018 16:29:03 +0200
Subject: [PATCH] PMCPESYS-76: cpufreq: Add Falcon Mtn plat to cpufreq-dt
 driver

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/cpufreq/cpufreq-dt-platdev.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/cpufreq/cpufreq-dt-platdev.c b/drivers/cpufreq/cpufreq-dt-platdev.c
index 71267626456b..2b7727ad926c 100644
--- a/drivers/cpufreq/cpufreq-dt-platdev.c
+++ b/drivers/cpufreq/cpufreq-dt-platdev.c
@@ -28,6 +28,8 @@ static const struct of_device_id machines[] __initconst = {
 
 	{ .compatible = "hisilicon,hi6220", },
 
+	{ .compatible = "intel,falconmx", },
+
 	{ .compatible = "fsl,imx27", },
 	{ .compatible = "fsl,imx51", },
 	{ .compatible = "fsl,imx53", },
