From b61fd54657b633681a0439327cdfc658d0ebca9c Mon Sep 17 00:00:00 2001
From: Rekha Eswaran <rekha.eswaran@intel.com>
Date: Fri, 24 Aug 2018 17:40:13 +0800
Subject: [PATCH] PONRTSYS-1908: remove qos_mib proc from Datapath

---
 .../lantiq/datapath/gswip31/datapath_proc.c        | 88 ----------------------
 1 file changed, 88 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_proc.c b/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_proc.c
index 84222478fbd2..5a95f029d164 100644
--- a/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_proc.c
+++ b/drivers/net/ethernet/lantiq/datapath/gswip31/datapath_proc.c
@@ -33,7 +33,6 @@
 #define PROC_PMAC  "pmac"
 #define PROC_EP "ep" /*EP/port ID info */
 #define DP_PROC_CBMLOOKUP "lookup"
-#define DP_MIB_Q "qos_mib"
 
 struct list_head fdb_tbl_list;
 
@@ -996,92 +995,6 @@ int get_p_mib(int inst, int pid,
 	PR_INFO("Q[%03d]:0x%08x 0x%08x 0x%08x 0x%08x\n",\
 		 i, c, x, y, z)\
 
-ssize_t proc_qos_mib(struct file *file, const char *buf,
-		     size_t count, loff_t *ppos)
-{
-	int len = 0;
-	char data[100];
-	char *param_list[10];
-	int num;
-	int start, end, i;
-	int inst = 0;
-	u32 c;
-	u32 total_accept_pkt, total_drop_pkt, red_drop_ktp;
-	u32 gree_b, yellow_b;
-
-	len = (count >= sizeof(data)) ? (sizeof(data) - 1) : count;
-	DP_DEBUG(DP_DBG_FLAG_DBG, "len=%d\n", len);
-	if (len <= 0) {
-		PR_ERR("Wrong len value (%d)\n", len);
-		return count;
-	}
-	if (copy_from_user(data, buf, len)) {
-		PR_ERR("copy_from_user fail");
-		return count;
-	}
-	data[len - 1] = 0; /* Make string */
-	num = dp_split_buffer(data, param_list, ARRAY_SIZE(param_list));
-	if (num < 1)
-		goto help;
-	if (dp_strncmpi(param_list[0], "q", 1) == 0) {
-		start = 0;
-		end = MAX_QUEUE;
-		if (param_list[1])
-			start = dp_atoi(param_list[1]);
-		if (param_list[2])
-			end = dp_atoi(param_list[2]);
-		if (start < 0)
-			start = 0;
-		if (end <= start)
-			end = start + 1;
-		if (end > MAX_QUEUE)
-			end = MAX_QUEUE;
-		PR_INFO("%5s:%10s %10s  %10s  %10s (%d-%d)\n",
-			"QID", "qocc(b)", "accept(p)", "drop(p)", "red_drop(p)",
-			start, end);
-		for (i = start; i < end; i++) {
-			get_q_qocc(inst, i, &c);
-			get_q_mib(inst, i, &total_accept_pkt,
-				  &total_drop_pkt, &red_drop_ktp);
-			if (c || total_accept_pkt || total_drop_pkt ||
-			    red_drop_ktp)
-				PRINT_Q_MIB(i, c, total_accept_pkt,
-					    total_drop_pkt, red_drop_ktp);
-		}
-	} else if (dp_strncmpi(param_list[0], "p", 1) == 0) {
-		start = 0;
-		end = MAX_CQM_DEQ;
-		if (param_list[1])
-			start = dp_atoi(param_list[1]);
-		if (param_list[2])
-			end = dp_atoi(param_list[2]);
-		if (start < 0)
-			start = 0;
-		if (end <= start)
-			end = start + 1;
-		if (end > MAX_CQM_DEQ)
-			end = MAX_CQM_DEQ;
-		PR_INFO("Port Id :green(b)   yellow(b)  (%d_%d)\n",
-			start, end);
-		for (i = start; i < end; i++) {
-			get_p_mib(inst, i, &gree_b, &yellow_b);
-			if (gree_b || yellow_b)
-				PR_INFO("P[%03d]: 0x%08x 0x%08x\n",
-					i, gree_b, yellow_b);
-		}
-	} else {
-		goto help;
-	}
-
-	return count;
-help:   /*                        [0]    [1]*/
-	PR_INFO("queue mib: echo <q> <start qid> <end qid> > %s\n",
-		"/proc/dp/" DP_MIB_Q);
-	PR_INFO("port  mib: echo <p> <start port_id> <end port_id> > %s\n",
-		"/proc/dp/" DP_MIB_Q);
-	return count;
-}
-
 static int proc_gsw_pce_dump(struct seq_file *s, int pos)
 {
 	struct core_ops *gsw_handle;
@@ -2490,7 +2403,6 @@ static struct dp_proc_entry dp_proc_entries[] = {
 #endif
 	{DP_PROC_CBMLOOKUP, NULL, lookup_dump31, lookup_start31,
 		proc_get_qid_via_index31},
-	{DP_MIB_Q, NULL, NULL, NULL, proc_qos_mib},
 
 	/*the last place holder */
 	{NULL, NULL, NULL, NULL, NULL}
