From 89b12afbef6dde3ff323b816e2c8ad1412d06fcb Mon Sep 17 00:00:00 2001
From: eugenehu <eugene.huang@intel.com>
Date: Mon, 20 Aug 2018 17:49:25 +0800
Subject: [PATCH] enable Fibocom L830-EB module

---
 drivers/net/usb/cdc_ncm.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/usb/cdc_ncm.c b/drivers/net/usb/cdc_ncm.c
index 3f9d6610f168..5b880b3da379 100644
--- a/drivers/net/usb/cdc_ncm.c
+++ b/drivers/net/usb/cdc_ncm.c
@@ -1798,6 +1798,13 @@ static const struct usb_device_id cdc_devs[] = {
 	  .driver_info = (unsigned long)&wwan_noarp_info,
 	},
 
+	/* Fibocom L830-EB LTE Modem platform */
+	{ USB_DEVICE_AND_INTERFACE_INFO(0x2cb7, 0x000b,
+		USB_CLASS_COMM,
+		USB_CDC_SUBCLASS_NCM, USB_CDC_PROTO_NONE),
+	  .driver_info = (unsigned long)&wwan_noarp_info,
+	},
+
 	/* Generic CDC-NCM devices */
 	{ USB_INTERFACE_INFO(USB_CLASS_COMM,
 		USB_CDC_SUBCLASS_NCM, USB_CDC_PROTO_NONE),
