From 44f9f969ebee114a7ebe200d51636e7d1651116a Mon Sep 17 00:00:00 2001
From: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
Date: Mon, 24 Sep 2018 18:26:29 +0200
Subject: [PATCH] PMCPESYS-89: spdmon: Replace panic() with dev_err()

Signed-off-by: Waldemar Rymarkiewicz <waldemarx.rymarkiewicz@intel.com>
---
 drivers/hwmon/spdmon.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/hwmon/spdmon.c b/drivers/hwmon/spdmon.c
index 33846701a96e..119a931068ea 100644
--- a/drivers/hwmon/spdmon.c
+++ b/drivers/hwmon/spdmon.c
@@ -826,13 +826,17 @@ static int spd_mon_probe(struct platform_device *pdev)
 	memset(pctrl, 0, sizeof(ltq_spd_mon_ctrl));
 
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
-	if (!res)
-		panic("Failed to get spd_mon resources\n");
+	if (!res) {
+		dev_err(&pdev->dev, "Failed to get spd_mon resources\n");
+		return -EINVAL;
+	};
 
 	pctrl->phybase = res->start;
 	pctrl->membase = devm_ioremap_resource(&pdev->dev, res);
-	if (!pctrl->membase)
-		panic("Failed to remap spd_mon resources\n");
+	if (IS_ERR(pctrl->membase)) {
+		dev_err(&pdev->dev, "Failed to remap spd_mon resources\n");
+		return PTR_ERR(pctrl->membase);
+	}
 
 	dev_info(&pdev->dev, "base address: 0x%x\t", (u32)pctrl->membase);
 	dev_info(&pdev->dev, "PHY base address: 0x%x\n", pctrl->phybase);
