From d395786bbad783692a82be7e2ce7337d79ee73e3 Mon Sep 17 00:00:00 2001
From: vijaysix <vijayx.singh@intel.com>
Date: Fri, 5 Oct 2018 14:31:48 +0530
Subject: [PATCH] Kernel config macro to enable downstream MPE accel feature
 for QOS

---
 net/Kconfig | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/net/Kconfig b/net/Kconfig
index 12ec833811d3..93e124fc4651 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -110,6 +110,14 @@ config INTEL_IPQOS_ACCEL_DISABLE
 
 	  If unsure  say N
 
+config INTEL_IPQOS_MPE_DS_ACCEL
+	bool "Support for DS Ingress QOS with MPE Acceleration"
+	depends on INTEL_IPQOS
+	help
+	  QOS Downstream MPE Acceleration Support
+
+	  If unsure, say N.
+
 config VLAN_8021Q_COPY_TO_EXTMARK
 	bool "Will copy 8021Q vlan-id and vlan-prio into extmark"
 	depends on VLAN_8021Q && NETWORK_EXTMARK
