From 98e77e3af19b813f8eb9017579e567ea7afaee06 Mon Sep 17 00:00:00 2001
From: Subhra Banerjee <subhrax.banerjee@intel.com>
Date: Thu, 11 Oct 2018 20:19:18 +0530
Subject: [PATCH] UGW_SW-31143: Compilation fixes for PPPoA

---
 net/atm/pppoatm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/atm/pppoatm.c b/net/atm/pppoatm.c
index 95ea49c5db8b..a71435088996 100644
--- a/net/atm/pppoatm.c
+++ b/net/atm/pppoatm.c
@@ -95,7 +95,7 @@ enum {
 	PPA_PPPOA_CHECK_IFACE,
 };
 
-extern int32_t ppa_get_pppoa_info_fn(struct net_device *dev,
+extern int32_t (*ppa_get_pppoa_info_fn)(struct net_device *dev,
 		void *pvcc, uint32_t pppoa_id, void *value);
 
 static int get_pppoa_info(struct net_device *dev, void *pvcc,
@@ -111,7 +111,7 @@ static int get_pppoa_info(struct net_device *dev, void *pvcc,
 		return -1;
 
 	switch (pppoa_id) {
-	case PPA_PPPOA_GET_VC:
+	case PPA_PPPOA_GET_VCC:
 		*patmvcc = p_atm_vcc->atmvcc;
 		break;
 	default:
