From 3efa6b10e1eaf46cb965494c3db500b2c550ab13 Mon Sep 17 00:00:00 2001
From: vijaysix <vijayx.singh@intel.com>
Date: Wed, 31 Oct 2018 10:39:49 +0530
Subject: [PATCH] UGW_SW-31838:Macro to set 16 priority in extmark for 8.x

---
 include/linux/extmark.h | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/linux/extmark.h b/include/linux/extmark.h
index 789c45c56d24..dd0d1a1d070a 100644
--- a/include/linux/extmark.h
+++ b/include/linux/extmark.h
@@ -4,12 +4,12 @@
 /* Bit counting starts from Least significant bit, 0..31*/
 
 #define CPEID_START_BIT_POS 0
-#define CPEID_NO_OF_BITS 6
-#define CPEID_MASK 0x3f
+#define CPEID_NO_OF_BITS 5
+#define CPEID_MASK 0x1f
 
-#define QUEPRIO_START_BIT_POS 6
-#define QUEPRIO_NO_OF_BITS 4
-#define QUEPRIO_MASK 0x3c0
+#define QUEPRIO_START_BIT_POS 5
+#define QUEPRIO_NO_OF_BITS 5
+#define QUEPRIO_MASK 0x3e0
 
 #define SUBQUEUE_START_BIT_POS 10
 #define SUBQUEUE_NO_OF_BITS 3
