From 94bf8cfdb7e14344aa55ebeddd1e27ac5a144557 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Mon, 5 Nov 2018 16:01:15 +0100
Subject: [PATCH] DRVLIB_SW-938: Forward some error code to caller

---
 drivers/clk/intel/clk-prx300.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/intel/clk-prx300.c b/drivers/clk/intel/clk-prx300.c
index 5a8e4add6346..f903a03f0d4c 100644
--- a/drivers/clk/intel/clk-prx300.c
+++ b/drivers/clk/intel/clk-prx300.c
@@ -304,11 +304,11 @@ static int __init intel_prx300_cgu_probe(struct platform_device *pdev)
 
 	map = syscon_node_to_regmap(np);
 	if (IS_ERR(map))
-		return -ENODEV;
+		return PTR_ERR(map);
 
 	ctx = intel_clk_init(map, PRX300_CLK_NR_CLKS);
 	if (IS_ERR(ctx))
-		return -ENOMEM;
+		return PTR_ERR(ctx);
 
 	ctx->np = np;
 	ctx->dev = dev;
