From 27ca878b54cca1654535f25427813bb8af1d3856 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Thu, 27 Dec 2018 15:58:16 +0800
Subject: [PATCH] DRVLIB_SW-1043: clear LD bit for LJPLL

---
 drivers/clk/intel/clk-cgu-pll.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/clk/intel/clk-cgu-pll.c b/drivers/clk/intel/clk-cgu-pll.c
index da44422eae73..4e2ea5cb615c 100644
--- a/drivers/clk/intel/clk-cgu-pll.c
+++ b/drivers/clk/intel/clk-cgu-pll.c
@@ -245,6 +245,13 @@ prx300_pll_set_params(struct intel_clk_pll *pll, unsigned int mult,
 	intel_set_clk_val(pll->map, pll->reg + 0x8, 18, 6, div);
 	intel_set_clk_val(pll->map, pll->reg, 2, 24, frac);
 	intel_set_clk_val(pll->map, pll->reg, 28, 1, !!frac);
+
+	if (pll->type == TYPE_LJPLL) {
+		intel_set_clk_val(pll->map, pll->reg, 31, 1, 1);
+		udelay(1);
+		intel_set_clk_val(pll->map, pll->reg, 31, 1, 0);
+	}
+
 	return intel_pll_wait_for_lock(pll);
 }
 
