From d542b81b5799be6c11c24b87ff6c10f634ac755d Mon Sep 17 00:00:00 2001
From: yixin zhu <yixin.zhu@intel.com>
Date: Tue, 19 Mar 2019 11:09:44 +0800
Subject: [PATCH] DRVLIB_SW-1602: fix slic clock get failed issue

---
 drivers/clk/intel/clk-prx300.c | 166 +++++++++++++++--------------------------
 1 file changed, 59 insertions(+), 107 deletions(-)

diff --git a/drivers/clk/intel/clk-prx300.c b/drivers/clk/intel/clk-prx300.c
index 9f94c8dc9a93..b88ad8133a45 100644
--- a/drivers/clk/intel/clk-prx300.c
+++ b/drivers/clk/intel/clk-prx300.c
@@ -18,73 +18,21 @@
 #define PLL_DIV_WIDTH		4
 #define PLL_DDIV_WIDTH		3
 
-/* Gate0 clock shift */
-#define G_XBAR0_SHIFT		0
-#define G_XBAR1_SHIFT		1
-#define G_XBAR7_SHIFT		7
-
-/* Gate1 clock shift */
-#define G_V_CODEC_SHIFT		2
-#define G_DMA0_SHIFT		3
-#define G_I2C0_SHIFT		4
-#define G_I2C1_SHIFT		5
-#define G_I2C2_SHIFT		6
-#define G_SPI1_SHIFT		7
-#define G_SPI0_SHIFT		8
-#define G_QSPI_SHIFT		9
-#define G_CQEM_SHIFT		10
-#define G_SSO_SHIFT			11
-#define G_GPTC0_SHIFT		12
-#define G_GPTC1_SHIFT		13
-#define G_GPTC2_SHIFT		14
-
-#define G_URT0_SHIFT		17
-#define G_URT1_SHIFT		18
-
-#define G_SECPT_SHIFT		21
-#define G_SCPU_SHIFT		22
-#define G_MPE_SHIFT			23
-
-#define G_TDM_SHIFT			25
-#define G_PP_SHIFT			26
-#define G_DMA3_SHIFT		27
-#define G_SWITCH_SHIFT		28
-#define G_PON_SHIFT			29
-#define G_AON_SHIFT			30
-#define G_DDR_SHIFT			31
-
-/* Gate2 clock shift */
-#define G_CTRL0_SHIFT		1
-#define G_MSI0_SHIFT		2
-#define G_PD0_SHIFT			7
-#define G_CTRL1_SHIFT		17
-#define G_MSI1_SHIFT		18
-#define G_PD1_SHIFT			23
-#define G_ASPI_SHIFT		28
-#define G_ADMA_SHIFT		29
-#define G_AHIF_SHIFT		30
-#define G_ASL_SHIFT			31
-
-/* Gate3 clock shift */
-#define GCLK_SWREF_SHIFT	4
-#define GCLK_CBPHY0_SHIFT	24
-#define GCLK_CBPHY1_SHIFT	25
-
 /* Register definition */
-#define PLL0A_CFG0	0x4
-#define PLL0B_CFG0	0x24
-#define PLL1_CFG0	0x44
-#define PLL2_CFG0	0x64
-#define LJPLL3_CFG0	0x84
-#define LJPLL4_CFG0	0xA4
-#define LJPLL5_CFG0	0xC4
-#define CGU_GATE0	0x114
-#define CGU_GATE1	0x120
-#define CGU_GATE2	0x130
-#define CGU_IF_CLK	0x140
-
-#define PLL_DIV(x)	((x) + 0x4)
-#define PLL_SSC(x)	((x) + 0x10)
+#define PLL0A_CFG0		0x4
+#define PLL0B_CFG0		0x24
+#define PLL1_CFG0		0x44
+#define PLL2_CFG0		0x64
+#define LJPLL3_CFG0		0x84
+#define LJPLL4_CFG0		0xA4
+#define LJPLL5_CFG0		0xC4
+#define CGU_GATE0		0x114
+#define CGU_GATE1		0x120
+#define CGU_GATE2		0x130
+#define CGU_IF_CLK		0x140
+
+#define PLL_DIV(x)		((x) + 0x4)
+#define PLL_SSC(x)		((x) + 0x10)
 
 static const struct clk_div_table pll_div[] = {
 	{1,	2},
@@ -190,93 +138,97 @@ static const struct intel_clk_branch prx300_branch_clks[] __initconst = {
 
 	/* Gate0 clocks */
 	INTEL_GATE(PRX300_GCLK_XBAR0, "g_xbar0", NULL, CLK_IGNORE_UNUSED,
-		   CGU_GATE0, G_XBAR0_SHIFT, GATE_CLK_HW, 0),
+		   CGU_GATE0, 0, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_XBAR1, "g_xbar1", NULL, CLK_IGNORE_UNUSED,
-		   CGU_GATE0, G_XBAR1_SHIFT, GATE_CLK_HW, 0),
+		   CGU_GATE0, 1, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_XBAR7, "g_xbar7", NULL, CLK_IGNORE_UNUSED,
-		   CGU_GATE0, G_XBAR7_SHIFT, GATE_CLK_HW, 0),
+		   CGU_GATE0, 7, GATE_CLK_HW, 0),
 
 	/* Gate1 clocks */
 	INTEL_GATE(PRX300_GCLK_V_CODEC, "g_code", NULL, 0, CGU_GATE1,
-		   G_V_CODEC_SHIFT, GATE_CLK_HW, 0),
+		   2, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_DMA0, "g_dma0", NULL, 0, CGU_GATE1,
-		   G_DMA0_SHIFT, GATE_CLK_HW, 0),
+		   3, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_I2C0, "g_i2c0", NULL, 0, CGU_GATE1,
-		   G_I2C0_SHIFT, GATE_CLK_HW, 0),
+		   4, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_I2C1, "g_i2c1", NULL, 0, CGU_GATE1,
-		   G_I2C1_SHIFT, GATE_CLK_HW, 0),
+		   5, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_I2C2, "g_i2c2", NULL, 0, CGU_GATE1,
-		   G_I2C2_SHIFT, GATE_CLK_HW, 0),
+		   6, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SPI1, "g_spi1", NULL, 0, CGU_GATE1,
-		   G_SPI1_SHIFT, GATE_CLK_HW, 0),
+		   7, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SPI0, "g_spi0", NULL, 0, CGU_GATE1,
-		   G_SPI0_SHIFT, GATE_CLK_HW, 0),
+		   8, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_QSPI, "g_qspi", NULL, 0, CGU_GATE1,
-		   G_QSPI_SHIFT, GATE_CLK_HW, 0),
+		   9, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_CQEM, "g_cqem", NULL, 0, CGU_GATE1,
-		   G_CQEM_SHIFT, GATE_CLK_HW, 0),
+		   10, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SSO, "g_sso", NULL, 0, CGU_GATE1,
-		   G_SSO_SHIFT, GATE_CLK_HW, 0),
+		   11, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_GPTC0, "g_gptc0", NULL, 0, CGU_GATE1,
-		   G_GPTC0_SHIFT, GATE_CLK_HW, 0),
+		   12, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_GPTC1, "g_gptc1", NULL, 0, CGU_GATE1,
-		   G_GPTC1_SHIFT, GATE_CLK_HW, 0),
+		   13, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_GPTC2, "g_gptc2", NULL, 0, CGU_GATE1,
-		   G_GPTC2_SHIFT, GATE_CLK_HW, 0),
+		   14, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_URT0, "g_uart0", NULL, 0, CGU_GATE1,
-		   G_URT0_SHIFT, GATE_CLK_HW, 0),
+		   17, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_URT1, "g_uart1", NULL, 0, CGU_GATE1,
-		   G_URT1_SHIFT, GATE_CLK_HW, 0),
+		   18, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SECPT, "g_secpt", NULL, 0,   CGU_GATE1,
-		   G_SECPT_SHIFT, GATE_CLK_HW, 0),
+		   21, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SCPU, "g_scpu", NULL, 0, CGU_GATE1,
-		   G_SCPU_SHIFT, GATE_CLK_HW, 0),
+		   22, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_MPE, "g_mpe", NULL, 0, CGU_GATE1,
-		   G_MPE_SHIFT, GATE_CLK_HW, 0),
+		   23, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_TDM, "g_tdm", NULL, 0, CGU_GATE1,
-		   G_TDM_SHIFT, GATE_CLK_HW, 0),
+		   25, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_PP, "g_pp", NULL, 0, CGU_GATE1,
-		   G_PP_SHIFT, GATE_CLK_HW, 0),
+		   26, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_DMA3, "g_dma3", NULL, 0, CGU_GATE1,
-		   G_DMA3_SHIFT, GATE_CLK_HW, 0),
+		   27, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_SWITCH, "g_switch", NULL, 0, CGU_GATE1,
-		   G_SWITCH_SHIFT, GATE_CLK_HW, 0),
+		   28, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_PON, "g_pon", "pondef", 0, CGU_GATE1,
-		   G_PON_SHIFT, GATE_CLK_HW, 0),
+		   29, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_AON, "g_aon", NULL, 0, CGU_GATE1,
-		   G_AON_SHIFT, GATE_CLK_HW, 0),
+		   30, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_DDR, "g_ddr", NULL, CLK_IGNORE_UNUSED, CGU_GATE1,
-		   G_DDR_SHIFT, GATE_CLK_HW, 0),
+		   31, GATE_CLK_HW, 0),
 
 	/* Gate2 clock */
 	INTEL_GATE(PRX300_GCLK_PCIE_CTRL0, "g_ctrl0", NULL, 0, CGU_GATE2,
-		   G_CTRL0_SHIFT, GATE_CLK_HW, 0),
+		   1, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_MSI0, "g_msi0", NULL, 0, CGU_GATE2,
-		   G_MSI0_SHIFT, GATE_CLK_HW, 0),
+		   2, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_PD0, "g_pd0", NULL, 0, CGU_GATE2,
-		   G_PD0_SHIFT, GATE_CLK_HW, 0),
+		   7, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_PCIE_CTRL1, "g_ctrl1", NULL, 0, CGU_GATE2,
-		   G_CTRL1_SHIFT, GATE_CLK_HW, 0),
+		   17, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_MSI1, "g_msi1", NULL, 0, CGU_GATE2,
-		   G_MSI1_SHIFT, GATE_CLK_HW, 0),
+		   18, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_PD1, "g_pd1", NULL, 0, CGU_GATE2,
-		   G_PD1_SHIFT, GATE_CLK_HW, 0),
+		   23, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_ASPI, "g_aspi", NULL, 0, CGU_GATE2,
-		   G_ASPI_SHIFT, GATE_CLK_HW, 0),
+		   28, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_ADMA, "g_adma", NULL, 0, CGU_GATE2,
-		   G_ADMA_SHIFT, GATE_CLK_HW, 0),
+		   29, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_AHIF, "g_ashif", NULL, 0, CGU_GATE2,
-		   G_AHIF_SHIFT, GATE_CLK_HW, 0),
+		   30, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_ASL, "g_asl", NULL, 0, CGU_GATE2,
-		   G_ASL_SHIFT, GATE_CLK_HW, 0),
+		   31, GATE_CLK_HW, 0),
 
 	/* Gate3 clock */
 	INTEL_GATE(PRX300_GCLK_SWREF, "g_swref", NULL, 0, CGU_IF_CLK,
-		   GCLK_SWREF_SHIFT, GATE_CLK_HW, 0),
+		   4, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_CBPHY0, "cbphy0", NULL, 0, CGU_IF_CLK,
-		   GCLK_CBPHY0_SHIFT, GATE_CLK_SW, 0),
+		   24, GATE_CLK_SW, 0),
 	INTEL_GATE(PRX300_GCLK_CBPHY1, "cbphy1", NULL, 0, CGU_IF_CLK,
-		   GCLK_CBPHY1_SHIFT, GATE_CLK_SW, 0)
+		   25, GATE_CLK_SW, 0),
+
+	/* Miscellaneous clocks */
+	INTEL_FIXED(PRX300_CLK_SLIC, "slic", NULL, 0, CGU_IF_CLK,
+		    14, 2, CLOCK_FLAG_VAL_INIT, 8192000, 2),
 };
 
 static const struct intel_clk_ddiv_data prx300_ddiv_clks[] __initconst = {
