From 6eda83653c8be4d2415965f2a2c060913f9ebd2e Mon Sep 17 00:00:00 2001
From: Thomas Bartholomae <t.bartholomae@intel.com>
Date: Thu, 28 Mar 2019 17:03:16 +0100
Subject: [PATCH] read additional properties from DT (config, calibration)

---
 drivers/hwmon/ina2xx.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/hwmon/ina2xx.c b/drivers/hwmon/ina2xx.c
index 1f291b344178..d47a701d1fc3 100644
--- a/drivers/hwmon/ina2xx.c
+++ b/drivers/hwmon/ina2xx.c
@@ -103,7 +103,7 @@ struct ina2xx_config {
 };
 
 struct ina2xx_data {
-	const struct ina2xx_config *config;
+	struct ina2xx_config *config;
 
 	long rshunt;
 	long current_lsb_uA;
@@ -114,7 +114,7 @@ struct ina2xx_data {
 	const struct attribute_group *groups[INA2XX_MAX_ATTRIBUTE_GROUPS];
 };
 
-static const struct ina2xx_config ina2xx_config[] = {
+static struct ina2xx_config ina2xx_config[] = {
 	[ina219] = {
 		.config_default = INA219_CONFIG_DEFAULT,
 		.calibration_value = 4096,
@@ -469,6 +469,12 @@ static int ina2xx_probe(struct i2c_client *client,
 
 	ina2xx_set_shunt(data, val);
 
+	if (of_property_read_u32(dev->of_node, "config", &val) >= 0)
+		data->config->config_default = val;
+
+	if (of_property_read_u32(dev->of_node, "calibration", &val) >= 0)
+		data->config->calibration_value = val;
+
 	ina2xx_regmap_config.max_register = data->config->registers;
 
 	data->regmap = devm_regmap_init_i2c(client, &ina2xx_regmap_config);
