From d932a04cd47506926fd1bc892ef0c8ae5fb7c45d Mon Sep 17 00:00:00 2001
From: "Gandhaveeti, Thathi ReddyX" <thathi.reddyx.gandhaveeti@intel.com>
Date: Wed, 10 Apr 2019 22:06:49 +0530
Subject: [PATCH] UGW_SW_37895, Addressed Review comments

---
 include/uapi/net/ppa_api.h | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index da452be0beac..467b460af43b 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -913,7 +913,7 @@ typedef struct {
 	\brief This is the data structure contains PPA session information.
  */
 typedef struct {
-	void* session; /*!< PPA SESSION pointer. Note, here we just use its address to delete a session for ioctl*/
+	void *session; /*!< PPA SESSION pointer. Note, here we just use its address to delete a session for ioctl*/
 	uint32_t hash;
 	uint16_t ip_proto; /*!< IP portocol TCP,UDP. */
 	uint16_t ip_tos; /*!< IP ToS value */
@@ -960,7 +960,7 @@ typedef struct {
 	\brief This is the data structure contains PPA session extra information.
  */
 typedef struct {
-	void* session; /*!< PPA SESSION pointer. Note, here we just use its address to modify a session for ioctl*/
+	void *session; /*!< PPA SESSION pointer. Note, here we just use its address to modify a session for ioctl*/
 	uint32_t hash; /* Hash value of the session */
 	PPA_SESSION_EXTRA session_extra; /*!< PPA SESSION extra pointer. */
 	uint32_t flags; /*!< Internal flag: PPA_F_SESSION_NEW_DSCP \n
@@ -974,7 +974,7 @@ typedef struct {
 	\brief This is the data structure for routing session Timer Info
  */
 typedef struct {
-	void* session; /*!< PPA SESSION pointer. Note, here we just use its address to modify a session for ioctl*/
+	void *session; /*!< PPA SESSION pointer. Note, here we just use its address to modify a session for ioctl*/
 	int32_t timer_in_sec; /*!< PPA SESSION polling timer in seconds. */
 	uint32_t flags; /*!< Reserved for future */
 } PPA_CMD_SESSION_TIMER;
@@ -1280,7 +1280,7 @@ typedef struct {
  */
 typedef struct EXP_FN_INFO {
 	uint8_t hookname[MAX_HOOK_NAME_LEN]; /*!< hook name */
-	void* hook_addr; /*!< mid hook address */
+	void *hook_addr; /*!< mid hook address */
 	uint8_t hook_flag:1; /*!< hook_flag: 0-disabled, otherwise -enabled */
 	uint8_t used_flag:1; /*!< used_flag: 0-not used, otherwise -used */
 	uint8_t reserved:6; /*!< reserved for future */
@@ -1531,7 +1531,7 @@ typedef struct {
 	\brief This is the data structure for SW session enabled info
  */
 typedef struct {
-	void* session; /*!< PPA SESSION pointer. Note, here we just use its address to enable a sw session for ioctl*/
+	void *session; /*!< PPA SESSION pointer. Note, here we just use its address to enable a sw session for ioctl*/
 	uint32_t sw_session_enable; /*!< enable/disable sw session enable flag */
 	uint32_t flags; /*!< reserved for future */
 } PPA_CMD_SW_SESSION_ENABLE_INFO;
