From a2a79433904864ecd912fd152cf35445346b72f0 Mon Sep 17 00:00:00 2001
From: adhisiva <dhivya.adhisivan@intel.com>
Date: Mon, 20 May 2019 20:25:58 +0530
Subject: [PATCH] UGW_SW-38795 : compilation failure fix when PPA disabled

---
 net/netfilter/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/net/netfilter/Kconfig b/net/netfilter/Kconfig
index 79030e5eb2be..75e76d29200e 100644
--- a/net/netfilter/Kconfig
+++ b/net/netfilter/Kconfig
@@ -972,6 +972,7 @@ config NETFILTER_XT_TARGET_SECMARK
 config PPA_IPTABLE_EVENT_HANDLING
 	tristate 'PPA iptable event handling'
 	default y
+	depends on PPA
 	help
 	  SDL Firewall Rule update for iptable with respect to PPA
 	  sessions programmed into the hardware
