From e5907119f0a976afe375349832839b7aca93542d Mon Sep 17 00:00:00 2001
From: "kamal.eradath" <kamal.eradath@intel.com>
Date: Wed, 19 Jun 2019 17:13:12 +0530
Subject: [PATCH] Added local session learning in capabilities.

---
 include/net/ppa/ppa_api.h | 13 ++-----------
 1 file changed, 2 insertions(+), 11 deletions(-)

diff --git a/include/net/ppa/ppa_api.h b/include/net/ppa/ppa_api.h
index 20545dc6eac0..ea9d7ab58efe 100644
--- a/include/net/ppa/ppa_api.h
+++ b/include/net/ppa/ppa_api.h
@@ -155,6 +155,8 @@ typedef enum {
 	SESS_IPV6,
 	SESS_MC_DS,
 	SESS_MC_DS_VAP,
+	SESS_LOCAL_IN,
+	SESS_LOCAL_OUT,
 	TUNNEL_6RD,
 	TUNNEL_DSLITE,
 	TUNNEL_L2TP_US,
@@ -920,17 +922,6 @@ typedef struct {
 	ppa_tnl_info	tunnel_info;	/*!< PPA tunnel info*/
 	void	 *hal_buffer;	/*!< PPA hal buffer*/
 } ppa_tunnel_entry;
-typedef struct{
-	uint16_t 	enabled; /*!< Entry is valid*/
-	uint16_t	f_ipv6;	/*!< PPA flag ipv6*/
-	IP_ADDR 	srcip;	/*!< PPA source ip*/
-	IP_ADDR 	dstip;	/*!< PPA destination ip*/
-	uint32_t 	srcport;	/*!< PPA source port*/
-	uint32_t 	dstport;	/*!< PPA destination port*/
-	uint16_t 	aggr_mtu;	/*!< PPA aggrate mtu*/
-	uint32_t 	timeout;	/*!< PPA timeout*/
-	uint16_t 	session_uid[2];	/*!< PPA session id*/
-} ppa_lro_entry;
 /*!
   \brief QoS Meter configuration structure
  */
