From e9305a099576d0279c913deb41bae5b3ffb5ce04 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Tue, 9 Jul 2019 11:22:50 +0800
Subject: [PATCH] DRVLIB_SW-2555: fix prx300 gptc compilation warning

---
 drivers/clocksource/intel-gptc-timer.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/clocksource/intel-gptc-timer.c b/drivers/clocksource/intel-gptc-timer.c
index dd06016d5831..04b485c57418 100644
--- a/drivers/clocksource/intel-gptc-timer.c
+++ b/drivers/clocksource/intel-gptc-timer.c
@@ -1204,7 +1204,9 @@ static int gptc_probe(struct platform_device *pdev)
 
 	gptc_of_init(np);
 
+#ifdef CONFIG_GPTC_SCHED_CLOCK
 	gptc_clocksource_init();
+#endif /* CONFIG_GPTC_SCHED_CLOCK */
 
 	/* Register immediately the clock event on BSP */
 	gptc_clkevent_init();
