From 7be288b894328690feaae4300cf79387c95ca60f Mon Sep 17 00:00:00 2001
From: Songjun Wu <songjun.wu@linux.intel.com>
Date: Fri, 20 Sep 2019 15:53:40 +0800
Subject: [PATCH] DRVLIB_SW-2849 - free clock

---
 drivers/clocksource/intel-gptc-timer.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/clocksource/intel-gptc-timer.c b/drivers/clocksource/intel-gptc-timer.c
index ddfdb6ad2d1d..3045bda60860 100644
--- a/drivers/clocksource/intel-gptc-timer.c
+++ b/drivers/clocksource/intel-gptc-timer.c
@@ -435,23 +435,29 @@ static int gptc_clock_init(struct gptc *gptc)
 	struct device_node *np = gptc->np;
 
 	gptc->gateclk = of_clk_get_by_name(np, "gptc");
-	if (IS_ERR_OR_NULL(gptc->gateclk)) {
+	if (IS_ERR(gptc->gateclk)) {
 		pr_err("Failed to get gptc gate clk: %ld\n",
 		       PTR_ERR(gptc->gateclk));
-		return gptc->gateclk ? PTR_ERR(gptc->gateclk) : -ENODEV;
+		return PTR_ERR(gptc->gateclk);
 	}
 
 	gptc->freqclk = of_clk_get_by_name(np, "freq");
-	if (IS_ERR_OR_NULL(gptc->freqclk)) {
+	if (IS_ERR(gptc->freqclk)) {
 		pr_err("Failed to get gptc frequency clk: %ld\n",
 		       PTR_ERR(gptc->freqclk));
-		return gptc->freqclk ? PTR_ERR(gptc->freqclk) : -ENODEV;
+		return PTR_ERR(gptc->freqclk);
 	}
 	return 0;
 }
 
 static void gptc_clock_deinit(struct gptc *gptc)
 {
+	if (gptc->gateclk && !IS_ERR(gptc->gateclk))
+		clk_put(gptc->gateclk);
+
+	if (gptc->freqclk && !IS_ERR(gptc->freqclk))
+		clk_put(gptc->freqclk);
+
 	gptc->gateclk = NULL;
 	gptc->freqclk = NULL;
 }
