From 9b16ecbf8a2c2ddb137d5bbf44f8937ef734e02b Mon Sep 17 00:00:00 2001
From: Subhra Banerjee <subhrax.banerjee@intel.com>
Date: Fri, 27 Sep 2019 16:48:42 +0530
Subject: [PATCH] LMTSYS-1398: Add WRR support in QoS FAPI.

---
 include/uapi/net/ppa_api.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index 517fcc65013c..94011191de82 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -1630,6 +1630,7 @@ typedef enum ppa_qos_module_type {
 typedef enum {
 	PPA_QOS_SCHED_SP = 0,
 	PPA_QOS_SCHED_WFQ = 1,
+	PPA_QOS_SCHED_WRR = 2,
 } PPA_QOS_QSCHED_MODE;
 /*!
 	\brief QoS Modify Queue Configuration structure
