From 8fecf3d69c3e474619d5d40d55c29b68b08ff072 Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Fri, 29 Nov 2019 17:14:57 +0800
Subject: [PATCH] DRV_DPM_SW-94: Umt Disable along with Umt release

---
 drivers/net/datapath/dpm/datapath_api.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index bfd337797fb3..a10ec941e3f4 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -967,6 +967,10 @@ static int remove_umt(int inst, const struct dp_umt_port *umt)
 	if (!ops)
 		return -ENODEV;
 
+	/* Disable UMT port */
+	if (ops->umt_enable(ops->umt_dev, umt->ctl.id, 0))
+		return DP_FAILURE;
+
 	return ops->umt_release(ops->umt_dev, umt->ctl.id);
 }
 #endif
