From 6f46a8415aa3ee4f0a2c5bb31bd00177ce7d9e5a Mon Sep 17 00:00:00 2001
From: Lishan Liu <lishan.liu@intel.com>
Date: Thu, 5 Dec 2019 15:51:55 +0800
Subject: [PATCH] add rst_cause info to rcu debugfs

---
 drivers/reset/reset-intel-syscon.c | 53 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 52 insertions(+), 1 deletion(-)

diff --git a/drivers/reset/reset-intel-syscon.c b/drivers/reset/reset-intel-syscon.c
index a88253b80908..9176400157e4 100644
--- a/drivers/reset/reset-intel-syscon.c
+++ b/drivers/reset/reset-intel-syscon.c
@@ -12,9 +12,12 @@
 #include <linux/reboot.h>
 #include <linux/regmap.h>
 #include <linux/reset-controller.h>
+#include <linux/seq_file.h>
+#include <linux/debugfs.h>
 
 #define RCU_RST_STAT		0x0024
 #define RCU_RST_REQ		0x0048
+#define STR_MAX_LEN		20
 
 /* reset platform data */
 #define to_reset_data(x)	container_of(x, struct intel_reset_data, rcdev)
@@ -31,8 +34,52 @@ struct intel_reset_data {
 	const struct intel_reset_soc_data *soc_data;
 	u32 *rst_ids;
 	u32 rst_num;
+	struct dentry *debugfs;
+	char reset_cause[STR_MAX_LEN];
 };
 
+static char rst_cause[STR_MAX_LEN] = "";
+
+static int __init get_reset_cause(char *line)
+{
+	strlcpy(rst_cause, line, STR_MAX_LEN - 1);
+	return 1;
+}
+
+__setup("rst_cause=", get_reset_cause);
+
+static int rst_cause_read(struct seq_file *s, void *v)
+{
+	struct intel_reset_data *rst_data = s->private;
+
+	seq_printf(s, "%s\n", rst_data->reset_cause);
+	return 0;
+}
+
+static int reset_debugfs_init(struct intel_reset_data *priv)
+{
+	struct dentry *file;
+	int ret;
+
+	priv->debugfs = debugfs_create_dir("reset", NULL);
+	if (!priv->debugfs)
+		return -ENOMEM;
+
+	file = debugfs_create_devm_seqfile(priv->dev, "rst_cause",
+					   priv->debugfs, rst_cause_read);
+	if (IS_ERR_OR_NULL(file)) {
+		ret = (int)PTR_ERR(file);
+		goto err;
+	}
+	return 0;
+
+err:
+	debugfs_remove_recursive(priv->debugfs);
+	priv->debugfs = NULL;
+	dev_err(priv->dev, "reset debugfs create fail!\n");
+	return ret;
+}
+
 static u32 intel_stat_reg_off(struct intel_reset_data *data, u32 req_off)
 {
 	if (data->soc_data->legacy && req_off == RCU_RST_REQ)
@@ -219,7 +266,11 @@ static int intel_reset_probe(struct platform_device *pdev)
 	if (err)
 		dev_warn(dev, "Failed to register restart handler\n");
 
-	return 0;
+	memcpy(data->reset_cause, rst_cause, STR_MAX_LEN);
+
+	err = reset_debugfs_init(data);
+
+	return err;
 }
 
 struct intel_reset_soc_data grx500_data = {
