From 7ca064ee8965631bb9f893da826aba6dce00fc1c Mon Sep 17 00:00:00 2001
From: Lishan Liu <lishan.liu@intel.com>
Date: Wed, 18 Dec 2019 11:29:56 +0800
Subject: [PATCH] resolve klocwork warning in umt driver

---
 drivers/net/datapath/umt/umt.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/net/datapath/umt/umt.c b/drivers/net/datapath/umt/umt.c
index 4ef5589919d8..8cdd71182654 100644
--- a/drivers/net/datapath/umt/umt.c
+++ b/drivers/net/datapath/umt/umt.c
@@ -645,8 +645,8 @@ static int umt_port_request(struct device *dev, struct dp_umt_port *port)
 {
 	struct umt_port *umt_p;
 	struct umt_priv *priv;
-	struct umt_port_ctl *ctl = &port->ctl;
-	struct umt_port_res *res = &port->res;
+	struct umt_port_ctl *ctl;
+	struct umt_port_res *res;
 	int id;
 
 	if (!port || !dev) {
@@ -655,6 +655,9 @@ static int umt_port_request(struct device *dev, struct dp_umt_port *port)
 	}
 	priv = dev_get_drvdata(dev);
 
+	ctl = &port->ctl;
+	res = &port->res;
+
 	id = umt_find_pid(priv, port);
 	if (id < 0) {
 		dev_err(dev, "Failed to find a suitable port ID\n");
@@ -805,12 +808,12 @@ static int umt_debugfs_init(struct umt_priv *priv)
 	if (!priv->debugfs)
 		return -ENOMEM;
 
-	file = debugfs_create_file("umt_info", 0644, priv->debugfs,
+	file = debugfs_create_file("umt_info", 0400, priv->debugfs,
 				   priv, &umt_port_fops);
 	if (!file)
 		goto err;
 
-	file = debugfs_create_file("umt_reg", 0644, priv->debugfs,
+	file = debugfs_create_file("umt_reg", 0400, priv->debugfs,
 				   priv, &umt_reg_info_fops);
 	if (!file)
 		goto err;
