From 414af436c3d0674d4d1e651a7e03108baeac8daf Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Fri, 3 Jan 2020 16:19:04 +0800
Subject: [PATCH] DRVLIB_SW-2847: resolve klockwork issue in CDMA driver

---
 drivers/dma/intel/cdma.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/dma/intel/cdma.c b/drivers/dma/intel/cdma.c
index 7a5026a4047d..a9764efce427 100755
--- a/drivers/dma/intel/cdma.c
+++ b/drivers/dma/intel/cdma.c
@@ -1461,7 +1461,7 @@ static int map_dma_chan(struct ifx_dma_chan_map *map, struct dma_ctrl *pctrl)
 
 	for (i = 0; i < dma_max_ports; i++)
 		strncpy(dma_devs[i].device_name,
-			dma_device_name[i], DMA_DEV_NAME_LEN);
+			dma_device_name[i], DMA_DEV_NAME_LEN - 1);
 
 	for (i = 0; i < pctrl->chans; i++) {
 		if (dma_chan_no_exist(i))
