From ac09f62a99d39e6d187acecb3ddbddddd34e4c82 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Tue, 7 Jan 2020 14:06:47 +0800
Subject: [PATCH] DRVLIB_SW-2847: resolve klockwork issue in CLK driver

---
 drivers/clk/intel/clk-prx300.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/clk/intel/clk-prx300.c b/drivers/clk/intel/clk-prx300.c
index 03b37ee397ae..e72a34cfa497 100644
--- a/drivers/clk/intel/clk-prx300.c
+++ b/drivers/clk/intel/clk-prx300.c
@@ -63,6 +63,7 @@ PNAME(pll_p)	= { "osc" };
 PNAME(emmc_p)	= { "emmc4", "noc4" };
 PNAME(sdio_p)	= { "sdio3", "sdio2" };
 
+static struct intel_clk_provider *early_ctx;
 static const struct intel_clk_early_data prx300_early_clks[] __initconst = {
 	{ PLL0A_CFG0, 0, 4, "earlycpu", pll_p, 0,
 	  0, 0, PLL_DIV_WIDTH,
@@ -261,27 +262,26 @@ static const struct intel_clk_ddiv_data prx300_ddiv_clks[] __initconst = {
 
 static void __init prx300_clk_init(struct device_node *np)
 {
-	struct intel_clk_provider *ctx;
 	struct regmap *map;
 
 	map = syscon_node_to_regmap(np);
 	if (IS_ERR(map))
 		return;
 
-	ctx = intel_clk_init(map, ARRAY_SIZE(prx300_early_clks));
+	early_ctx = intel_clk_init(map, ARRAY_SIZE(prx300_early_clks));
 
-	if (IS_ERR(ctx)) {
+	if (IS_ERR(early_ctx)) {
 		pr_err("%s: early clock alloc ctx failed!\n", __func__);
 		regmap_exit(map);
 		return;
 	}
 
-	if (intel_clk_register_early(ctx, prx300_early_clks,
+	if (intel_clk_register_early(early_ctx, prx300_early_clks,
 				     ARRAY_SIZE(prx300_early_clks))) {
-		kfree(ctx);
+		kfree(early_ctx);
 		return;
 	}
-	of_clk_add_hw_provider(np, of_clk_hw_onecell_get, &ctx->clk_data);
+	of_clk_add_hw_provider(np, of_clk_hw_onecell_get, &early_ctx->clk_data);
 }
 
 CLK_OF_DECLARE_DRIVER(intel_prx300_cgu, "intel,cgu-prx300",
