From 6bf2c16c932a3a34ed03ec17866d773547b741d1 Mon Sep 17 00:00:00 2001
From: leichuan <Chuanhua.lei@intel.com>
Date: Mon, 23 Dec 2019 18:15:43 +0800
Subject: [PATCH] Remove CMA 256MB limitation

---
 arch/mips/kernel/setup.c | 12 +-----------
 1 file changed, 1 insertion(+), 11 deletions(-)

diff --git a/arch/mips/kernel/setup.c b/arch/mips/kernel/setup.c
index b911c31fc89c..5c7c7fdb147f 100644
--- a/arch/mips/kernel/setup.c
+++ b/arch/mips/kernel/setup.c
@@ -776,7 +776,6 @@ static void __init request_crashkernel(struct resource *res)
 
 static void __init arch_mem_init(char **cmdline_p)
 {
-	phys_addr_t dma_cma_limit;
 	struct memblock_region *reg;
 	extern void plat_mem_setup(void);
 
@@ -871,16 +870,7 @@ static void __init arch_mem_init(char **cmdline_p)
 
 	plat_swiotlb_setup();
 
-	/*
-	 * dma_alloc_coherent will return uncached address for most legacy MIPS,
-	 * the physical RAM is 256 MB for interaptiv, physical RAM is up to 3GB,
-	 * however, the system has only 256MB uncached address available 
-	 */
-	dma_cma_limit = PHYS_OFFSET + SZ_256M;
-	if (PFN_PHYS(max_low_pfn) < dma_cma_limit)
-		dma_cma_limit = PFN_PHYS(max_low_pfn);
-
-	dma_contiguous_reserve(dma_cma_limit);
+	dma_contiguous_reserve(PFN_PHYS(max_low_pfn));
 	/* Tell bootmem about cma reserved memblock section */
 	for_each_memblock(reserved, reg)
 		if (reg->size != 0)
