From 976ad70aa83f1c5e13f6a6dde109e7ae9f1d27e0 Mon Sep 17 00:00:00 2001
From: yixin zhu <yixin.zhu@intel.com>
Date: Wed, 11 Mar 2020 11:55:20 +0800
Subject: [PATCH] DRVLIB_SW-3488: Fix UMT feature capability issue

---
 drivers/net/datapath/umt/umt.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/net/datapath/umt/umt.c b/drivers/net/datapath/umt/umt.c
index 8cdd71182654..7cff7ae69659 100644
--- a/drivers/net/datapath/umt/umt.c
+++ b/drivers/net/datapath/umt/umt.c
@@ -274,21 +274,21 @@ umt_check_cfg(struct umt_priv *priv, struct umt_port *umt_p,
 
 	/* check UMT RX CAP */
 	if (!(BIT(res->rx_src) & umt_p->rx_src_cap)) {
-		dev_dbg(dev, "UMT RX src: %u not supported!\n", res->rx_src);
+		dev_err(dev, "UMT RX src: %u not supported!\n", res->rx_src);
 		return -EINVAL;
 	}
 
 	/* check UMT DMA CAP */
 	if (res->rx_src == UMT_RX_SRC_DMA) {
 		if (res->dma_ch_num > umt_p->dma_chn_cap) {
-			dev_dbg(dev, "UMT RX DMA CHNUM: %u not supported on pid: %u\n",
+			dev_err(dev, "UMT RX DMA CHNUM: %u not supported on pid: %u\n",
 				res->dma_ch_num, umt_p->id);
 			return -EINVAL;
 		}
 
 		/* check UMT DMA Ctrl/channel CAP */
 		if (umt_dma_check(priv, res->dma_id)) {
-			dev_dbg(dev, "UMT RX DMA: %u not supported on pid: %u\n",
+			dev_err(dev, "UMT RX DMA: %u not supported on pid: %u\n",
 				res->dma_id, umt_p->id);
 			return -EINVAL;
 		}
@@ -296,7 +296,7 @@ umt_check_cfg(struct umt_priv *priv, struct umt_port *umt_p,
 
 	/* check UMT feature CAP */
 	if ((ctl->fflag & umt_p->feature_cap) != ctl->fflag) {
-		dev_dbg(dev,
+		dev_err(dev,
 			"UMT feature not supported: req:0x%lx, hwcap:0x%lx\n",
 			ctl->fflag, umt_p->feature_cap);
 		return -EINVAL;
@@ -637,7 +637,8 @@ static int umt_port_release(struct device *dev, unsigned int id)
  * API to request and allocate UMT port
  * input:
  * @dev: umt device
- * @umt: UMT port info. if id is 0xff means dynamic allocate port
+ * @umt: UMT port info.
+ * UMT always dynamic allocate port
  * output:
  * ret: Fail: < 0,  Success: 0
  */
@@ -860,7 +861,8 @@ static const struct umt_soc_data prx300_umt_data = {
 	.ver          = PRX300_SOC,
 	.umt_port_num = 3,
 	.rx_src_cap   = BIT(UMT_RX_SRC_DMA),
-	.feature_cap  = UMT_SND_DIS | UMT_CLEAR_CNT,
+	.feature_cap  = UMT_SND_DIS | UMT_CLEAR_CNT
+			| UMT_NOT_SND_ZERO_CNT | UMT_SND_RX_CNT_ONLY,
 	.dq_base      = 4,
 };
 
