From 85da79620a45d5f2887f4536d8a465513a32b49c Mon Sep 17 00:00:00 2001
From: "santosh.kumar.mamila" <santosh.kumarx.mamila@intel.com>
Date: Thu, 12 Mar 2020 18:12:56 +0800
Subject: [PATCH] DRVLIB_SW-3464 - Klockwork fixes for CPU Frequency driver.

---
 drivers/cpufreq/xrx500-cpufreq.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/cpufreq/xrx500-cpufreq.c b/drivers/cpufreq/xrx500-cpufreq.c
index 0060b517cfac..ba0aef15c75b 100644
--- a/drivers/cpufreq/xrx500-cpufreq.c
+++ b/drivers/cpufreq/xrx500-cpufreq.c
@@ -127,7 +127,8 @@ static unsigned int xrx500_pnum2version(struct device *dev, unsigned int id)
 
 int xrx500_opp_set_supported_hw(struct device *cpu_dev)
 {
-	unsigned int version, id;
+	unsigned int id;
+	int version;
 	int ret;
 
 	id = ltq_get_cpu_id();
