From af65605a68acf1c80cb64f56f3d36d5d2edd4fac Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Mon, 6 Apr 2020 16:42:59 +0800
Subject: [PATCH] DRVLIB_SW-3142: grx500 wdt timeout value not read from dts

---
 drivers/watchdog/grx500_wdt.c | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index a8a04a0f82ca..bb8b837b0514 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -424,6 +424,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 	for_each_online_cpu(cpu) {
 		uint32_t config0;
 		struct watchdog_device *grx500_wdt;
+		u32 timeout_init;
 
 		grx500_wdt = &per_cpu(grx500wdt, cpu);
 
@@ -437,10 +438,22 @@ static int grx500wdt_probe(struct platform_device *pdev)
 		grx500_wdt->max_hw_heartbeat_ms = grx500_wdt->max_timeout *1000;
 		grx500_wdt->min_hw_heartbeat_ms = grx500_wdt->min_timeout *1000;
 
-		timeout = 3;
-		watchdog_init_timeout(grx500_wdt, timeout, &pdev->dev);
+		timeout = 0;
+		ret = watchdog_init_timeout(grx500_wdt, timeout, &pdev->dev);
+		if (ret)
+			dev_warn(&pdev->dev, "[%s] Set to U32_MAX\n", __func__);
+
+		if (!ret && ((u64)cpu_clk * (u64)(grx500_wdt->timeout)) <= (u64)U32_MAX)
+			timeout_init = (u64)cpu_clk * (u64)(grx500_wdt->timeout);
+		else
+			timeout_init = U32_MAX;
+
 		watchdog_set_nowayout(grx500_wdt, nowayout);
 
+		dev_dbg(&pdev->dev, "[%s]:[%d] timeout=%u timeout_init %u\n",
+			  __func__, __LINE__, grx500_wdt->timeout,
+			 timeout_init);
+
 		if (cpu == smp_processor_id()) {
 			gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0), 0x0);
 			wmb();
@@ -450,7 +463,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 				(config0 | WD_TYPE_SCD | WD_NWAIT));
 			wmb();
 			gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_INITIAL0),
-				(U32_MAX));
+				(timeout_init));
 			wmb();
 		} else {
 			local_irq_save(flags);
@@ -464,7 +477,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 				(config0 | WD_TYPE_SCD | WD_NWAIT));
 			wmb();
 			gic_write_reg(GIC_REG(VPE_OTHER, GIC_VPE_WD_INITIAL0),
-				(U32_MAX));
+				(timeout_init));
 			wmb();
 			local_irq_restore(flags);
 		}
