From 741555e35d3e89b0de1112ee9da8f09285066510 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Mon, 6 Apr 2020 16:36:01 +0800
Subject: [PATCH] DRVLIB_SW-3142: grx500 wdt cpu id mapping is not correct

---
 drivers/watchdog/grx500_wdt.c | 55 +++++++++++++++++++++++++++++++------------
 1 file changed, 40 insertions(+), 15 deletions(-)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index bb8b837b0514..062ea4cfb917 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -135,10 +135,12 @@ static int grx500wdt_start(struct watchdog_device *wdt_dev)
 {
 	uint32_t config0;
 	unsigned long flags;
+	struct grx500_wdt_struct *priv;
 
-	pr_debug("[%s]:[%d] wdt_dev=0x%p id=%d cpu = %d\n", __func__, __LINE__,
-		wdt_dev, wdt_dev->id, smp_processor_id());
-	if (wdt_dev->id == smp_processor_id()) {
+	priv = watchdog_get_drvdata(wdt_dev);
+	pr_debug("[%s]:[%d] wdd=0x%p id=%d cpu=%d priv->cpu=%d\n", __func__,
+		 __LINE__, wdt_dev, wdt_dev->id, smp_processor_id(), priv->cpu);
+	if (priv->cpu == smp_processor_id()) {
 		config0 = gic_read_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0));
 		rmb();
 		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0),
@@ -146,7 +148,8 @@ static int grx500wdt_start(struct watchdog_device *wdt_dev)
 		wmb();
 	} else {
 		local_irq_save(flags);
-		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR), wdt_dev->id);
+		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR),
+			      priv->cpu);
 		wmb();
 		config0 = gic_read_reg(GIC_REG(VPE_OTHER, GIC_VPE_WD_CONFIG0));
 		rmb();
@@ -163,10 +166,13 @@ static int grx500wdt_stop(struct watchdog_device *wdt_dev)
 {
 	uint32_t config0;
 	unsigned long flags;
+	struct grx500_wdt_struct *priv;
+
+	priv = watchdog_get_drvdata(wdt_dev);
 
-	pr_debug("[%s]:[%d] wdt_dev=0x%p id=%d cpu = %d\n", __func__, __LINE__,
-		wdt_dev, wdt_dev->id, smp_processor_id());
-	if (wdt_dev->id == smp_processor_id()) {
+	pr_debug("[%s]:[%d] wdd=0x%p id=%d cpu=%d priv->cpu=%d\n", __func__,
+		 __LINE__, wdt_dev, wdt_dev->id, smp_processor_id(), priv->cpu);
+	if (priv->cpu == smp_processor_id()) {
 		config0 = gic_read_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0));
 		rmb();
 		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0),
@@ -174,7 +180,8 @@ static int grx500wdt_stop(struct watchdog_device *wdt_dev)
 		wmb();
 	} else {
 		local_irq_save(flags);
-		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR), wdt_dev->id);
+		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR),
+			      priv->cpu);
 		wmb();
 		config0 = gic_read_reg(GIC_REG(VPE_OTHER, GIC_VPE_WD_CONFIG0));
 		rmb();
@@ -192,6 +199,9 @@ static int grx500wdt_set_timeout(struct watchdog_device *wdt_dev,
 {
 	struct watchdog_device *grx500_wdt;
 	unsigned long flags;
+	struct grx500_wdt_struct *priv;
+
+	priv = watchdog_get_drvdata(wdt_dev);
 
 	if (!capable(CAP_SYS_ADMIN))
 		return -EPERM;
@@ -200,12 +210,13 @@ static int grx500wdt_set_timeout(struct watchdog_device *wdt_dev,
 	grx500_wdt = wdt_dev;
 
 	grx500_wdt->timeout = new_timeout;
-	pr_debug("%s: timeout = %d, cpu = %d, id = %d wdt_dev=0x%p\n", __func__,
-		new_timeout, smp_processor_id(), wdt_dev->id, wdt_dev);
+	pr_debug("%s: timeout=%d, cpu=%d, id=%d wdd=0x%p priv->cpu=%d\n",
+		 __func__, new_timeout, smp_processor_id(), wdt_dev->id,
+		 wdt_dev, priv->cpu);
 
 	grx500wdt_stop(grx500_wdt);
 
-	if (wdt_dev->id == smp_processor_id()) {
+	if (priv->cpu == smp_processor_id()) {
 		if (((u64)cpu_clk * (u64)(grx500_wdt->timeout)) <= (u64)U32_MAX) {
 			gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_INITIAL0),
 				(cpu_clk * grx500_wdt->timeout));
@@ -216,7 +227,8 @@ static int grx500wdt_set_timeout(struct watchdog_device *wdt_dev,
 		wmb();
 	} else {
 		local_irq_save(flags);
-		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR), wdt_dev->id);
+		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR),
+			      priv->cpu);
 		wmb();
 		if (((u64)cpu_clk * (u64)(grx500_wdt->timeout)) <= (u64)U32_MAX) {
 			gic_write_reg(GIC_REG(VPE_OTHER, GIC_VPE_WD_INITIAL0),
@@ -238,8 +250,11 @@ static uint32_t grx500wdt_get_timeleft(struct watchdog_device *wdt_dev)
 {
 	u32 count0, initial0, config0, rst_en, map0;
 	unsigned long flags;
+	struct grx500_wdt_struct *priv;
+
+	priv = watchdog_get_drvdata(wdt_dev);
 
-	if (wdt_dev->id == smp_processor_id()) {
+	if (priv->cpu == smp_processor_id()) {
 		initial0 = gic_read_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_INITIAL0));
 		rmb();
 		config0 = gic_read_reg(GIC_REG(VPE_LOCAL, GIC_VPE_WD_CONFIG0));
@@ -250,7 +265,8 @@ static uint32_t grx500wdt_get_timeleft(struct watchdog_device *wdt_dev)
 		rmb();
 	} else {
 		local_irq_save(flags);
-		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR), wdt_dev->id);
+		gic_write_reg(GIC_REG(VPE_LOCAL, GIC_VPE_OTHER_ADDR),
+			      priv->cpu);
 		wmb();
 		initial0 = gic_read_reg(GIC_REG(VPE_OTHER, GIC_VPE_WD_INITIAL0));
 		rmb();
@@ -404,13 +420,22 @@ static int grx500wdt_probe(struct platform_device *pdev)
 
 	for_each_possible_cpu(cpu) {
 		struct watchdog_device *grx500_wdt;
+		struct grx500_wdt_struct *priv_data;
 
 		grx500_wdt = &per_cpu(grx500wdt, cpu);
 
 		grx500_wdt->id = cpu;
 		grx500_wdt->info = &grx500wdt_info;
 		grx500_wdt->ops = &grx500wdt_ops;
-		watchdog_set_drvdata(grx500_wdt, priv);
+
+		priv_data = devm_kzalloc(&pdev->dev, sizeof(*priv_data),
+					 GFP_KERNEL);
+		if (!priv)
+			return -ENOMEM;
+		memcpy(priv_data, priv, sizeof(*priv));
+
+		priv_data->cpu = cpu;
+		watchdog_set_drvdata(grx500_wdt, priv_data);
 	}
 
 	regmap_read(ltq_rcu_base, RCU_IAP_WDT_RST_STAT, &resetcause);
