From 177ee20c0e18fdde83cbd668e034e89451db546a Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Thu, 23 Apr 2020 17:58:41 +0800
Subject: [PATCH] DRVLIB_SW-3142: grx500 wdt klocwork issues

---
 drivers/watchdog/grx500_wdt.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index 062ea4cfb917..beebd909895b 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -414,6 +414,12 @@ static int grx500wdt_probe(struct platform_device *pdev)
 	}
 
 	cpu_clk = clk_get_rate(clk);
+	if (!cpu_clk) {
+		dev_err(&pdev->dev, "Failed to get CPU clock rate\n");
+		ret = -EINVAL;
+		goto exit;
+	}
+
 	dev_info(&pdev->dev, "[%s]:[%d] cpu_clk=%lu\n",
 		__func__, __LINE__, cpu_clk);
 	/* cpu_clk = grx500_clk_get_rate((struct clk *)ltq_grx500_wdt->clk); */
@@ -430,7 +436,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 
 		priv_data = devm_kzalloc(&pdev->dev, sizeof(*priv_data),
 					 GFP_KERNEL);
-		if (!priv)
+		if (!priv_data)
 			return -ENOMEM;
 		memcpy(priv_data, priv, sizeof(*priv));
 
@@ -438,6 +444,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 		watchdog_set_drvdata(grx500_wdt, priv_data);
 	}
 
+	devm_kfree(&pdev->dev, priv);
 	regmap_read(ltq_rcu_base, RCU_IAP_WDT_RST_STAT, &resetcause);
 	dev_info(&pdev->dev, "[%s]WDT reset is Bit31, RCU_IAP_WDT_RST_STAT=0x%x\n",
 		__func__, resetcause);
