From 7e25ea87a669171c6d7ede33a3d4df6e201d4008 Mon Sep 17 00:00:00 2001
From: Jack Ping CHNG <jack.ping.chng@linux.intel.com>
Date: Fri, 5 Jun 2020 10:52:48 +0800
Subject: [PATCH] DRV_DPM_SW-291: DPM fix klocwork pointer dev check for null

---
 drivers/net/datapath/dpm/datapath_instance.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/datapath/dpm/datapath_instance.c b/drivers/net/datapath/dpm/datapath_instance.c
index d680103b90eb..cb2a75c56822 100644
--- a/drivers/net/datapath/dpm/datapath_instance.c
+++ b/drivers/net/datapath/dpm/datapath_instance.c
@@ -315,6 +315,9 @@ int dp_inst_del_dev(struct net_device *dev, char *subif_name, int inst,
 	int offset;
 #endif
 
+	if (!dev)
+		return DP_FAILURE;
+
 	dp_dev = dp_dev_lookup(dev);
 	if (!dp_dev) {
 		pr_err("Failed to dp_dev_lookup: %s_%s\n",
