From 1d50e35bf6e9ae9930840ea1ecb8b55c24c77864 Mon Sep 17 00:00:00 2001
From: skodalix <srinivasx.kodali@intel.com>
Date: Thu, 17 Sep 2020 11:08:54 +0530
Subject: [PATCH] UGW_SW-45166: make carrier OFF till the physical port
 registration

---
 net/atm/br2684.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/net/atm/br2684.c b/net/atm/br2684.c
index 59e7ca395b29..74b684367378 100644
--- a/net/atm/br2684.c
+++ b/net/atm/br2684.c
@@ -771,6 +771,12 @@ static int br2684_create(void __user *arg)
 		free_netdev(netdev);
 		return err;
 	}
+
+	/* By default, carrier state is ON and it's causing in making the interface
+	 * UP even before physical port registration. Hence making the carrier OFF
+	 * and it will be ON after physical port registration */
+	netif_carrier_off(netdev);
+
 	/* Mark br2684 device */
 	netdev->priv_flags |= IFF_BR2684;
 
