From f28fe1ad08f4f9b7d097ec88b53ed3451d8a55de Mon Sep 17 00:00:00 2001
From: Deepansh Rastogi <deepansh.rastogi@intel.com>
Date: Tue, 29 Sep 2020 21:53:57 +0530
Subject: [PATCH] UGW_SW-53878: Passing hal id during PPA HW meter add

---
 include/net/ppa/ppa_api.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/net/ppa/ppa_api.h b/include/net/ppa/ppa_api.h
index 6deb012418b5..753554ad31f2 100644
--- a/include/net/ppa/ppa_api.h
+++ b/include/net/ppa/ppa_api.h
@@ -1003,7 +1003,7 @@ int32_t ppa_add_class_rule(PPA_CLASS_RULE *rule);
 int32_t ppa_mod_class_rule(PPA_CLASS_RULE *rule);
 int32_t ppa_del_class_rule(PPA_CLASS_RULE *rule);
 int32_t ppa_get_class_rule(PPA_CLASS_RULE *rule);
-int32_t ppa_set_qos_meter(PPA_QOS_METER_CFG *meter_cfg);
+int32_t ppa_set_qos_meter(PPA_QOS_METER_CFG *meter_cfg, uint32_t hal_id);
 #endif
 int32_t ppa_inactivity_status(PPA_U_SESSION *);
 int32_t ppa_set_session_inactivity(PPA_U_SESSION *, int32_t);
