From cd149aa21377fffe3bafe4cdbf5e5ea39aa6d01b Mon Sep 17 00:00:00 2001
From: agaraix <anath.bandux.garai@intel.com>
Date: Sun, 25 Oct 2020 00:43:25 +0530
Subject: [PATCH] UGW_SW-54868: ip6_tunnel: Removed imbalance rcu_read_unlock()

---
 net/ipv6/ip6_tunnel.c | 34 ++++++++++++++++------------------
 1 file changed, 16 insertions(+), 18 deletions(-)

diff --git a/net/ipv6/ip6_tunnel.c b/net/ipv6/ip6_tunnel.c
index c17b5cd87ac6..7413bfeda59c 100644
--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -956,24 +956,22 @@ static int __ip6_tnl_rcv(struct ip6_tnl *tunnel, struct sk_buff *skb,
 	memset(skb->cb, 0, sizeof(struct inet6_skb_parm));
 
 	if (tpi->proto == htons(ETH_P_IP) && tunnel->parms.fmrs &&
-		!ipv6_addr_equal(&ipv6h->saddr, &tunnel->parms.raddr)) {
-			/* Packet didn't come from BR, so lookup FMR */
-			struct __ip6_tnl_fmr *fmr;
-			struct in6_addr expected = tunnel->parms.raddr;
-			for (fmr = tunnel->parms.fmrs; fmr; fmr = fmr->next)
-				if (ipv6_prefix_equal(&ipv6h->saddr,
-					&fmr->ip6_prefix, fmr->ip6_prefix_len))
-						break;
-
-			/* Check that IPv6 matches IPv4 source to prevent spoofing */
-			if (fmr)
-				ip4ip6_fmr_calc(&expected, ip_hdr(skb),
-						skb_tail_pointer(skb), fmr, false);
-
-			if (!ipv6_addr_equal(&ipv6h->saddr, &expected)) {
-				rcu_read_unlock();
-				goto drop;
-			}
+	    !ipv6_addr_equal(&ipv6h->saddr, &tunnel->parms.raddr)) {
+		/* Packet didn't come from BR, so lookup FMR */
+		struct __ip6_tnl_fmr *fmr;
+		struct in6_addr expected = tunnel->parms.raddr;
+		for (fmr = tunnel->parms.fmrs; fmr; fmr = fmr->next)
+			if (ipv6_prefix_equal(&ipv6h->saddr,
+			    &fmr->ip6_prefix, fmr->ip6_prefix_len))
+				break;
+
+		/* Check that IPv6 matches IPv4 source to prevent spoofing */
+		if (fmr)
+			ip4ip6_fmr_calc(&expected, ip_hdr(skb),
+					skb_tail_pointer(skb), fmr, false);
+
+		if (!ipv6_addr_equal(&ipv6h->saddr, &expected))
+			goto drop;
 	}
 
 	__skb_tunnel_rx(skb, tunnel->dev, tunnel->net);
