From 81406483e2f6a0409663a664ff8e9579f8d69e95 Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Thu, 28 Jan 2021 12:14:13 +0800
Subject: [PATCH] PONRTSYS-8497: Warning fix in DPM

---
 drivers/net/datapath/dpm/datapath_pce.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/net/datapath/dpm/datapath_pce.c b/drivers/net/datapath/dpm/datapath_pce.c
index c643452afdd9..4e9d2c5ba13a 100644
--- a/drivers/net/datapath/dpm/datapath_pce.c
+++ b/drivers/net/datapath/dpm/datapath_pce.c
@@ -572,7 +572,6 @@ static int pce_rule_del(GSW_PCE_ruleDelete_t *pce,
 			struct dp_subblk_entry *subblk_entry)
 {
 	struct core_ops *ops = gsw_get_swcore_ops(0);
-	GSW_PCE_GlobalBitMap_t gbitmap = {0};
 
 	if (!ops)
 		return -1;
