From aac3b9fa25740f59863e3ffece2cdbeb9d9fb11f Mon Sep 17 00:00:00 2001
From: Anath Bandhu Garai <anath.bandux.garai@intel.com>
Date: Wed, 5 May 2021 20:36:41 +0530
Subject: [PATCH] PUMARTSYS-42620: dcdp api for CGR triband configuration

---
 include/net/directconnect_dp_api.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/net/directconnect_dp_api.h b/include/net/directconnect_dp_api.h
index 7ea1c6b1c3ed..f538e2f68959 100644
--- a/include/net/directconnect_dp_api.h
+++ b/include/net/directconnect_dp_api.h
@@ -52,7 +52,7 @@
 /**
   \brief DC DP API version code
  */
-#define DC_DP_API_VERSION_CODE        0x050004
+#define DC_DP_API_VERSION_CODE        0x050005
 
 /**
   \brief DC DP API version
@@ -625,6 +625,7 @@ enum dc_dp_dev_type {
 	DC_DP_DEV_WAV_500 = 0,
 	DC_DP_DEV_WAV_600_24G,
 	DC_DP_DEV_WAV_600_5G,
+	DC_DP_DEV_WAV_600_6G,
 	DC_DP_DEV_WAV_600_CDB,
 };
 
