From bee812c8e5a43d8a67ce7da1bfeca203dd7efad6 Mon Sep 17 00:00:00 2001
From: Anath Bandhu Garai <anath.bandux.garai@intel.com>
Date: Wed, 5 May 2021 20:40:58 +0530
Subject: [PATCH] UGW_SW-57825 : LGM: dcdp api for fw download

---
 include/net/directconnect_dp_api.h | 69 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 69 insertions(+)

diff --git a/include/net/directconnect_dp_api.h b/include/net/directconnect_dp_api.h
index 4c082b728c69..4c89157599e9 100644
--- a/include/net/directconnect_dp_api.h
+++ b/include/net/directconnect_dp_api.h
@@ -654,6 +654,19 @@ enum dc_dp_power_state {
 	DC_DP_PS_BOOST,
 };
 
+/** \brief FW Download options for FW allocation and release.
+ */
+enum dc_dp_fw_opt_type {
+	/* Free struct firmware and buffer (if alloc’ed) before return */
+	DC_DP_FW_OPT_FREE_BUF_AFTER_CB  = BIT(0),
+	/* Allocate FW uncached non-coherent buffer */
+	DC_DP_FW_OPT_MEM_ALLOC_UC       = BIT(1),
+	/* Allocate FW coherent buffer */
+	DC_DP_FW_OPT_MEM_ALLOC_COHERENT = BIT(2),
+	/* For debug builds, allow rewriting firmware */
+	DC_DP_FW_OPT_DEBUG_RELOAD       = BIT(3),
+};
+
 /* @} */
 
 /** \addtogroup DirectConnect_Datapath_Lib_Structs */
@@ -1218,12 +1231,68 @@ struct dc_dp_prop {
 	uint8_t dgaf_disabled:1; /*!< [in, out] Setting for DGAF Disable: 0 or 1 */
 };
 
+/** \brief Forward declaration of Linux structure
+ */
+struct firmware;
+struct device;
 /* @} */
 
 
 /** \addtogroup DirectConnect_Datapath_Lib_APIs */
 /* @{ */
 
+/** \brief FW download callback – Can be used to program HW block
+ \param[in] device: device for which firmware is being loaded
+ \param[in] fw_p: pointer to firmware image
+ \param[in] dma_addr:  Phys addr of the FW download buffer
+ \return 0 if OK / -1 if error
+ \note This callback is optional – driver may use this to program HW block
+ */
+typedef int (*dc_dp_soc_fw_load_to_hw)(
+	struct device *device,
+	struct firmware *fw_p,
+	dma_addr_t dma_addr
+); /*!< FW Download function callback */
+
+/** \brief FW download and NOC firewall abstraction
+  \param[in] firmware_p : pointer to firmware image
+  \param[in] name: name of firmware file
+  \param[in] device: device for which firmware is being loaded
+  \param[in] cb: Optional download callback
+  \param[in] opt_type: FW download options for FW/Buffer alloc/free
+  \return 0 if OK / -1 if error
+  \note Expect FW in contiguous DDR, since is to be used by SoC HW
+    blocks. ROData can also be allocated behind the code being
+    downloaded for use by the HW block.
+*/
+int32_t
+dc_dp_soc_request_firmware_load (
+	const struct firmware **firmware_p,
+	char *name,
+	struct device *device,
+	dc_dp_soc_fw_load_to_hw cb,
+	enum dc_dp_fw_opt_type type
+);
+
+/** \brief FW download and NOC firewall abstraction
+  \param[in] firmware_p : pointer to firmware image
+  \param[in] name: name of firmware file
+  \param[in] device: device for which firmware is being loaded
+  \param[in] opt_type: FW download options for FW/Buffer alloc/free
+  \return 0 if OK / -1 if error
+  \note Shall remove NoC Firewall rule (if needed)
+    If Free flag is passed, allocated buffer shall be released
+    (only if allocated by dc_dp_soc_request_firmware_load())
+    Will release struct firmware and members
+*/
+void
+dc_dp_soc_release_firmware (
+	const struct firmware **firmware_p,
+	const char *name,
+	struct device *device,
+	enum dc_dp_fw_opt_type type
+);
+
 /** \brief  Obtain device capability in SoC.
   \param[in] dev_id  enum indicating device type
   \param[out] dev_config  Device configuration
