From 79b2b3285a5af69b0c208d3090c8fd489a68ccb9 Mon Sep 17 00:00:00 2001
From: "Shao, Guohua" <gshao@maxlinear.com>
Date: Tue, 8 Jun 2021 20:06:44 +0800
Subject: [PATCH] DRV_DPM_SW-501 temp solution to reduce stack usage in dpm/ppa

---
 include/net/datapath_api.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index c7de71348806..82892357f904 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -266,7 +266,7 @@ enum DP_IO_PORT_TYPE {
 #define DP_MAX_CPU 4 /*!< @brief Max number of CPUs */
 #define DP_PMAP_PCP_NUM 8  /*!< @brief  Max pcp entries supported per pmapper*/
 #define DP_PMAP_DSCP_NUM 64 /*!<@brief  Max dscp entries supported per pmapper*/
-#define DP_MAX_CTP_PER_DEV  64  /*!< @brief  max CTP per dev:
+#define DP_MAX_CTP_PER_DEV  8   /*!< @brief  max CTP per dev:
 				 *  Note: its value should be like
 				 *     max(DP_PMAP_DSCP_NUM, DP_PMAP_PCP_NUM)
 				 *   ? Maybe we can reduce from 64 to 8 since
