From a523633c1fade468b16d579445b7f046e8c854e5 Mon Sep 17 00:00:00 2001
From: "Shao, Guohua" <gshao@maxlinear.com>
Date: Thu, 15 Jul 2021 10:13:52 +0800
Subject: [PATCH] DRV_DPM_SW-522: dpm fix some Kconfig dependency issue

---
 drivers/net/datapath/dpm/Kconfig | 28 ++++++----------------------
 1 file changed, 6 insertions(+), 22 deletions(-)

diff --git a/drivers/net/datapath/dpm/Kconfig b/drivers/net/datapath/dpm/Kconfig
index 96c71c65d020..52095765f12b 100644
--- a/drivers/net/datapath/dpm/Kconfig
+++ b/drivers/net/datapath/dpm/Kconfig
@@ -23,7 +23,6 @@ config INTEL_DATAPATH_PARTIAL_FEEDS
 
 config INTEL_DATAPATH_ACA_CSUM_WORKAROUND
 	bool "ACA Checksum Workaround"
-	default n
 	depends on GRX500_CBM && INTEL_DATAPATH
 	---help---
 	  It is to solve system bus hang hang issue in GRX500
@@ -43,7 +42,6 @@ config INTEL_DATAPATH_MANUAL_PARSE
 
 config INTEL_DATAPATH_COPY_LINEAR_BUF_ONLY
 	bool "Datapath Copy linear buffer only for skb"
-	default n
 	depends on GRX500_CBM && INTEL_DATAPATH
 	---help---
 	  Datapath Copy linear buffer only for skb if need to alloc new buffer.
@@ -63,8 +61,7 @@ config INTEL_DATAPATH_DBG
 
 config INTEL_DATAPATH_DBG_PROTOCOL_PARSE
 	bool "Datapath Debug Tool for hw checksum's protocol parsing"
-	default n
-	depends on GRX500_CBM && INTEL_DATAPATH_DBG
+	depends on INTEL_DATAPATH_DBG
 	---help---
 	  Datapath Debug Tool for hw checksum's protocol parsing
 	  Only for debugging purpose
@@ -72,7 +69,6 @@ config INTEL_DATAPATH_DBG_PROTOCOL_PARSE
 
 config INTEL_DATAPATH_EXTRA_DEBUG
 	bool "extra debugging support"
-	default n
 	depends on  INTEL_DATAPATH_DBG
 	---help---
 	  This is to enable/disable extra strict debugging support.
@@ -82,16 +78,14 @@ config INTEL_DATAPATH_EXTRA_DEBUG
 
 config INTEL_DATAPATH_SKB
 	bool "Datapath Skb Hack"
-	default n
-	depends on INTEL_DATAPATH
+	depends on GRX500_CBM && INTEL_DATAPATH
 	---help---
 	  For Ethernet OAM and MPE FW purpose testing purpose,
 	  It needs to hack SKB
 
 config INTEL_DATAPATH_ETH_OAM
 	bool "ETH OAM SUPPORT"
-	default n
-	depends on  GRX500_CBM && INTEL_DATAPATH_SKB
+	depends on GRX500_CBM && INTEL_DATAPATH_SKB
 	---help---
 	  Datapath Ethernet OAM support. Once it is enabled, it will add some fields in skb structure
 	  in order to support MPE FAST HOOK. The reason is that some network driver is
@@ -100,7 +94,6 @@ config INTEL_DATAPATH_ETH_OAM
 
 config INTEL_DATAPATH_SWITCHDEV
 	bool "Switchdev Support"
-	default n
 	depends on  INTEL_DATAPATH && NET_SWITCHDEV && !GRX500_CBM
 	---help---
 	  Switchdev support for different switch in datapath
@@ -108,7 +101,6 @@ config INTEL_DATAPATH_SWITCHDEV
 config INTEL_DATAPATH_CPUFREQ
 	bool "Datapath DFS(COC) support"
 	depends on GRX500_CBM && INTEL_DATAPATH && CPU_FREQ
-	default n
 	---help---
 	  It is to support DFS(COC) in Datapath
 
@@ -137,23 +129,15 @@ config LTQ_DATAPATH_CPUFREQ
 	bool
 	default INTEL_DATAPATH_CPUFREQ
 
-menuconfig INTEL_DATAPATH_HAL_GSWIP31
-	bool "Datapath HAL_GSWIP31"
+config INTEL_DATAPATH_HAL_GSWIP31
+	bool "Datapath HAL"
 	default y
 	depends on INTEL_DATAPATH && PRX300_CQM
-	---help---
-	  Datapath Lib is to provide common rx/tx wrapper Lib without taking
-	  care of much HW knowledge and also provide common interface for legacy
-	  devices and different HW like to CBM or LRO.
-	  Take note: All devices need to register to datapath Lib first
 
-if INTEL_DATAPATH_HAL_GSWIP31
 config INTEL_DATAPATH_QOS_HAL
-	bool "datapath QOS hal"
+	bool "Datapath HAL-QOS"
 	default y
 	depends on PPV4
-endif
 
 source "drivers/net/datapath/dpm/gswip30/Kconfig"
 endif
-
