--- a/include/fapi_sys_common.h
+++ b/include/fapi_sys_common.h
@@ -29,9 +29,14 @@
 #include <net/ppa_api.h>
 #endif
 #ifdef CONFIG_LANTIQ_SWITCH
+#ifdef PLATFORM_XRX750
+#include <net/switch-api/lantiq_gsw.h>
+#include <net/switch-api/lantiq_gsw_api.h>
+#else
 #include <net/switch_api/lantiq_gsw.h>
 #include <net/switch_api/lantiq_gsw_api.h>
 #endif
+#endif
 #include "fapi_processorstat.h"
 #include "fapi_led.h"
 
