# HG changeset patch
# Parent  067d18fd3316741c15137420025b4a50a4cf0998

diff --git a/cmd/mxl/atom_ia_switch.c b/cmd/mxl/atom_ia_switch.c
new file mode 100644
--- /dev/null
+++ b/cmd/mxl/atom_ia_switch.c
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright © 2022 MaxLinear, Inc.
+ *
+ */
+#include <common.h>
+#include <command.h>
+#include <image.h>
+
+extern void ia_vmm(void);
+
+int do_ia_vmm(struct cmd_tbl *cmdtp, int flag, int argc, char * const argv[])
+{
+	printf("Changing IA_BOOT to IA_VMM ..");
+	ia_vmm();
+	printf(" done\n");
+	return 0;
+}
+
+U_BOOT_CMD(
+	ia_switch, CONFIG_SYS_MAXARGS, 0, do_ia_vmm,
+	"change IA_BOOT SAI to IA_VMM SAI",
+	"\t allows X86 to change from IA_BOOT to IA_VMM\n"
+);
